Binary files a/.cache/clangd/index/anon_file.c.99B9E6E03541840D.idx and b/.cache/clangd/index/anon_file.c.99B9E6E03541840D.idx differ
Binary files a/.cache/clangd/index/anon_file.h.BB3645A747F6D424.idx and b/.cache/clangd/index/anon_file.h.BB3645A747F6D424.idx differ
Binary files a/.cache/clangd/index/bitscan.c.9E27A99F16ED1A27.idx and b/.cache/clangd/index/bitscan.c.9E27A99F16ED1A27.idx differ
Binary files a/.cache/clangd/index/bitscan.h.C9065F8E86577802.idx and b/.cache/clangd/index/bitscan.h.C9065F8E86577802.idx differ
Binary files a/.cache/clangd/index/c11_compat.h.25A44B3026F71602.idx and b/.cache/clangd/index/c11_compat.h.25A44B3026F71602.idx differ
Binary files a/.cache/clangd/index/c99_compat.h.3BDFB279B6BCB4FB.idx and b/.cache/clangd/index/c99_compat.h.3BDFB279B6BCB4FB.idx differ
Binary files a/.cache/clangd/index/compiler.h.3927B1896D392069.idx and b/.cache/clangd/index/compiler.h.3927B1896D392069.idx differ
Binary files a/.cache/clangd/index/cso_cache.c.14D925DA80BE5458.idx and b/.cache/clangd/index/cso_cache.c.14D925DA80BE5458.idx differ
Binary files a/.cache/clangd/index/cso_cache.h.5C49BCB99C87DBD5.idx and b/.cache/clangd/index/cso_cache.h.5C49BCB99C87DBD5.idx differ
Binary files a/.cache/clangd/index/cso_hash.c.3323B6245F906EFC.idx and b/.cache/clangd/index/cso_hash.c.3323B6245F906EFC.idx differ
Binary files a/.cache/clangd/index/cso_hash.h.31CC94CABF231173.idx and b/.cache/clangd/index/cso_hash.h.31CC94CABF231173.idx differ
Binary files a/.cache/clangd/index/detect_os.h.288882ECDBFC6981.idx and b/.cache/clangd/index/detect_os.h.288882ECDBFC6981.idx differ
Binary files a/.cache/clangd/index/drm_hw.h.DF95F496C026E9E6.idx and b/.cache/clangd/index/drm_hw.h.DF95F496C026E9E6.idx differ
Binary files a/.cache/clangd/index/drm_renderer.h.BA691FE2E4B87C4A.idx and b/.cache/clangd/index/drm_renderer.h.BA691FE2E4B87C4A.idx differ
Binary files a/.cache/clangd/index/fast_urem_by_const.h.DA3749F4E1B05BE7.idx and b/.cache/clangd/index/fast_urem_by_const.h.DA3749F4E1B05BE7.idx differ
Binary files a/.cache/clangd/index/hash_table.c.C02D52976775D7E0.idx and b/.cache/clangd/index/hash_table.c.C02D52976775D7E0.idx differ
Binary files a/.cache/clangd/index/hash_table.h.5C0AAC003E45B2DF.idx and b/.cache/clangd/index/hash_table.h.5C0AAC003E45B2DF.idx differ
Binary files a/.cache/clangd/index/iov.c.0D7F23DDB696B00C.idx and b/.cache/clangd/index/iov.c.0D7F23DDB696B00C.idx differ
Binary files a/.cache/clangd/index/list.h.B04F1AD8580165D1.idx and b/.cache/clangd/index/list.h.B04F1AD8580165D1.idx differ
Binary files a/.cache/clangd/index/macros.h.B9032FD945249669.idx and b/.cache/clangd/index/macros.h.B9032FD945249669.idx differ
Binary files a/.cache/clangd/index/no_extern_c.h.DF9A4C85FAEA1AD9.idx and b/.cache/clangd/index/no_extern_c.h.DF9A4C85FAEA1AD9.idx differ
Binary files a/.cache/clangd/index/os_file.c.D6A362E80173EBCA.idx and b/.cache/clangd/index/os_file.c.D6A362E80173EBCA.idx differ
Binary files a/.cache/clangd/index/os_file.h.695B679026CF9E5A.idx and b/.cache/clangd/index/os_file.h.695B679026CF9E5A.idx differ
Binary files a/.cache/clangd/index/os_memory_aligned.h.26CC3D893135F89D.idx and b/.cache/clangd/index/os_memory_aligned.h.26CC3D893135F89D.idx differ
Binary files a/.cache/clangd/index/os_memory_stdc.h.727D1BBC2BE7EE25.idx and b/.cache/clangd/index/os_memory_stdc.h.727D1BBC2BE7EE25.idx differ
Binary files a/.cache/clangd/index/os_memory.h.A9D56BA02F81348C.idx and b/.cache/clangd/index/os_memory.h.A9D56BA02F81348C.idx differ
Binary files a/.cache/clangd/index/os_misc.c.9DDD1F9AEFA44F10.idx and b/.cache/clangd/index/os_misc.c.9DDD1F9AEFA44F10.idx differ
Binary files a/.cache/clangd/index/os_misc.h.722F693467F3E7B3.idx and b/.cache/clangd/index/os_misc.h.722F693467F3E7B3.idx differ
Binary files a/.cache/clangd/index/p_compiler.h.2AE8F9B9ED3BFED4.idx and b/.cache/clangd/index/p_compiler.h.2AE8F9B9ED3BFED4.idx differ
Binary files a/.cache/clangd/index/p_config.h.FFF06000E63EB2B4.idx and b/.cache/clangd/index/p_config.h.FFF06000E63EB2B4.idx differ
Binary files a/.cache/clangd/index/p_defines.h.E7C6984C3B6633A8.idx and b/.cache/clangd/index/p_defines.h.E7C6984C3B6633A8.idx differ
Binary files a/.cache/clangd/index/p_format.h.33053DF6A89751A5.idx and b/.cache/clangd/index/p_format.h.33053DF6A89751A5.idx differ
Binary files a/.cache/clangd/index/p_shader_tokens.h.EF283D325C1976B4.idx and b/.cache/clangd/index/p_shader_tokens.h.EF283D325C1976B4.idx differ
Binary files a/.cache/clangd/index/p_state.h.43171AFE6E620D3D.idx and b/.cache/clangd/index/p_state.h.43171AFE6E620D3D.idx differ
Binary files a/.cache/clangd/index/proxy_renderer.h.BC0F4D4902EC0337.idx and b/.cache/clangd/index/proxy_renderer.h.BC0F4D4902EC0337.idx differ
Binary files a/.cache/clangd/index/ralloc.c.FCA057FCF8AB68A5.idx and b/.cache/clangd/index/ralloc.c.FCA057FCF8AB68A5.idx differ
Binary files a/.cache/clangd/index/ralloc.h.02A485F5A4FB0925.idx and b/.cache/clangd/index/ralloc.h.02A485F5A4FB0925.idx differ
Binary files a/.cache/clangd/index/tgsi_build.c.8A4AFA66D093E8C1.idx and b/.cache/clangd/index/tgsi_build.c.8A4AFA66D093E8C1.idx differ
Binary files a/.cache/clangd/index/tgsi_build.h.7936FF4E486FBCC0.idx and b/.cache/clangd/index/tgsi_build.h.7936FF4E486FBCC0.idx differ
Binary files a/.cache/clangd/index/tgsi_dump.c.6B20D06C72699478.idx and b/.cache/clangd/index/tgsi_dump.c.6B20D06C72699478.idx differ
Binary files a/.cache/clangd/index/tgsi_dump.h.D69C06E108183519.idx and b/.cache/clangd/index/tgsi_dump.h.D69C06E108183519.idx differ
Binary files a/.cache/clangd/index/tgsi_info.c.C67E264B6EA6C5D2.idx and b/.cache/clangd/index/tgsi_info.c.C67E264B6EA6C5D2.idx differ
Binary files a/.cache/clangd/index/tgsi_info.h.AC97D233162FE48F.idx and b/.cache/clangd/index/tgsi_info.h.AC97D233162FE48F.idx differ
Binary files a/.cache/clangd/index/tgsi_iterate.c.E569D24487C69914.idx and b/.cache/clangd/index/tgsi_iterate.c.E569D24487C69914.idx differ
Binary files a/.cache/clangd/index/tgsi_iterate.h.D350C807C9D1A9C2.idx and b/.cache/clangd/index/tgsi_iterate.h.D350C807C9D1A9C2.idx differ
Binary files a/.cache/clangd/index/tgsi_parse.c.7DE0CC9EC63FFB94.idx and b/.cache/clangd/index/tgsi_parse.c.7DE0CC9EC63FFB94.idx differ
Binary files a/.cache/clangd/index/tgsi_parse.h.7CF75B4CF0FA4B06.idx and b/.cache/clangd/index/tgsi_parse.h.7CF75B4CF0FA4B06.idx differ
Binary files a/.cache/clangd/index/tgsi_sanity.c.13B3BE6664659060.idx and b/.cache/clangd/index/tgsi_sanity.c.13B3BE6664659060.idx differ
Binary files a/.cache/clangd/index/tgsi_sanity.h.263049DC8635609C.idx and b/.cache/clangd/index/tgsi_sanity.h.263049DC8635609C.idx differ
Binary files a/.cache/clangd/index/tgsi_scan.c.C8269FCDE010BB96.idx and b/.cache/clangd/index/tgsi_scan.c.C8269FCDE010BB96.idx differ
Binary files a/.cache/clangd/index/tgsi_scan.h.DC6EC20904AB2894.idx and b/.cache/clangd/index/tgsi_scan.h.DC6EC20904AB2894.idx differ
Binary files a/.cache/clangd/index/tgsi_strings.c.6915C65209B81C23.idx and b/.cache/clangd/index/tgsi_strings.c.6915C65209B81C23.idx differ
Binary files a/.cache/clangd/index/tgsi_strings.h.A744CB7E5E16AEDE.idx and b/.cache/clangd/index/tgsi_strings.h.A744CB7E5E16AEDE.idx differ
Binary files a/.cache/clangd/index/tgsi_text.c.A97C9246E902C6BA.idx and b/.cache/clangd/index/tgsi_text.c.A97C9246E902C6BA.idx differ
Binary files a/.cache/clangd/index/tgsi_text.h.411A9BB9D6727D09.idx and b/.cache/clangd/index/tgsi_text.h.411A9BB9D6727D09.idx differ
Binary files a/.cache/clangd/index/tgsi_util.c.EFBF0B02051D6FD0.idx and b/.cache/clangd/index/tgsi_util.c.EFBF0B02051D6FD0.idx differ
Binary files a/.cache/clangd/index/tgsi_util.h.7576B1F39F37BE66.idx and b/.cache/clangd/index/tgsi_util.h.7576B1F39F37BE66.idx differ
Binary files a/.cache/clangd/index/threads_posix.h.E03D440F81EC6052.idx and b/.cache/clangd/index/threads_posix.h.E03D440F81EC6052.idx differ
Binary files a/.cache/clangd/index/threads.h.9D162BFD7A4AB497.idx and b/.cache/clangd/index/threads.h.9D162BFD7A4AB497.idx differ
Binary files a/.cache/clangd/index/u_atomic.h.7AEEE844EA71D1AB.idx and b/.cache/clangd/index/u_atomic.h.7AEEE844EA71D1AB.idx differ
Binary files a/.cache/clangd/index/u_cpu_detect.c.897C1AFA2F2CBAC4.idx and b/.cache/clangd/index/u_cpu_detect.c.897C1AFA2F2CBAC4.idx differ
Binary files a/.cache/clangd/index/u_cpu_detect.h.B9903064BEC2C6F8.idx and b/.cache/clangd/index/u_cpu_detect.h.B9903064BEC2C6F8.idx differ
Binary files a/.cache/clangd/index/u_debug_describe.c.5AB78EBD44611692.idx and b/.cache/clangd/index/u_debug_describe.c.5AB78EBD44611692.idx differ
Binary files a/.cache/clangd/index/u_debug_describe.h.3377900B85E76739.idx and b/.cache/clangd/index/u_debug_describe.h.3377900B85E76739.idx differ
Binary files a/.cache/clangd/index/u_debug_refcnt.h.B2E8D50CCC1924B0.idx and b/.cache/clangd/index/u_debug_refcnt.h.B2E8D50CCC1924B0.idx differ
Binary files a/.cache/clangd/index/u_debug.c.9E855CF55C040F11.idx and b/.cache/clangd/index/u_debug.c.9E855CF55C040F11.idx differ
Binary files a/.cache/clangd/index/u_debug.h.3FFD92CDB8533EA9.idx and b/.cache/clangd/index/u_debug.h.3FFD92CDB8533EA9.idx differ
Binary files a/.cache/clangd/index/u_double_list.h.0F4F049A3CB97E31.idx and b/.cache/clangd/index/u_double_list.h.0F4F049A3CB97E31.idx differ
Binary files a/.cache/clangd/index/u_dual_blend.h.A14AAD3850C49AD3.idx and b/.cache/clangd/index/u_dual_blend.h.A14AAD3850C49AD3.idx differ
Binary files a/.cache/clangd/index/u_endian.h.8A8016836669F1D6.idx and b/.cache/clangd/index/u_endian.h.8A8016836669F1D6.idx differ
Binary files a/.cache/clangd/index/u_format_s3tc.h.5A92B1B6CE52222E.idx and b/.cache/clangd/index/u_format_s3tc.h.5A92B1B6CE52222E.idx differ
Binary files a/.cache/clangd/index/u_format.c.A306654B4E7F2266.idx and b/.cache/clangd/index/u_format.c.A306654B4E7F2266.idx differ
Binary files a/.cache/clangd/index/u_format.h.38E2AF6F959A62BD.idx and b/.cache/clangd/index/u_format.h.38E2AF6F959A62BD.idx differ
Binary files a/.cache/clangd/index/u_hash_table.c.0E93B4C69D86FB8C.idx and b/.cache/clangd/index/u_hash_table.c.0E93B4C69D86FB8C.idx differ
Binary files a/.cache/clangd/index/u_hash_table.h.0F8138D2075D0998.idx and b/.cache/clangd/index/u_hash_table.h.0F8138D2075D0998.idx differ
Binary files a/.cache/clangd/index/u_inlines.h.B389E4DD50433C8A.idx and b/.cache/clangd/index/u_inlines.h.B389E4DD50433C8A.idx differ
Binary files a/.cache/clangd/index/u_math.c.392417D40A413961.idx and b/.cache/clangd/index/u_math.c.392417D40A413961.idx differ
Binary files a/.cache/clangd/index/u_math.h.764D4A487004438C.idx and b/.cache/clangd/index/u_math.h.764D4A487004438C.idx differ
Binary files a/.cache/clangd/index/u_memory.h.456943C706481E44.idx and b/.cache/clangd/index/u_memory.h.456943C706481E44.idx differ
Binary files a/.cache/clangd/index/u_pointer.h.0487FFE4EAD8A4EA.idx and b/.cache/clangd/index/u_pointer.h.0487FFE4EAD8A4EA.idx differ
Binary files a/.cache/clangd/index/u_prim.h.30455C94ADAF901E.idx and b/.cache/clangd/index/u_prim.h.30455C94ADAF901E.idx differ
Binary files a/.cache/clangd/index/u_string.h.17FC1FDDA856B27E.idx and b/.cache/clangd/index/u_string.h.17FC1FDDA856B27E.idx differ
Binary files a/.cache/clangd/index/u_texture.c.06FFC3131216D38A.idx and b/.cache/clangd/index/u_texture.c.06FFC3131216D38A.idx differ
Binary files a/.cache/clangd/index/u_texture.h.5AFFC25086F85FFA.idx and b/.cache/clangd/index/u_texture.h.5AFFC25086F85FFA.idx differ
Binary files a/.cache/clangd/index/u_thread.h.608C3B49D511E2E7.idx and b/.cache/clangd/index/u_thread.h.608C3B49D511E2E7.idx differ
Binary files a/.cache/clangd/index/util.c.4E695E020D3F42EF.idx and b/.cache/clangd/index/util.c.4E695E020D3F42EF.idx differ
Binary files a/.cache/clangd/index/util.h.568F948E9E1C7613.idx and b/.cache/clangd/index/util.h.568F948E9E1C7613.idx differ
Binary files a/.cache/clangd/index/venus_context.c.8CB65E8EE8BE01DC.idx and b/.cache/clangd/index/venus_context.c.8CB65E8EE8BE01DC.idx differ
Binary files a/.cache/clangd/index/venus_context.h.46449588446EDFB6.idx and b/.cache/clangd/index/venus_context.h.46449588446EDFB6.idx differ
Binary files a/.cache/clangd/index/venus_hw.h.703C1A72CCDD01A2.idx and b/.cache/clangd/index/venus_hw.h.703C1A72CCDD01A2.idx differ
Binary files a/.cache/clangd/index/virgl_context.c.886857C7F88E55C6.idx and b/.cache/clangd/index/virgl_context.c.886857C7F88E55C6.idx differ
Binary files a/.cache/clangd/index/virgl_context.h.1D9C9327B811916A.idx and b/.cache/clangd/index/virgl_context.h.1D9C9327B811916A.idx differ
Binary files a/.cache/clangd/index/virgl_hw.h.97191BCB0E2552C8.idx and b/.cache/clangd/index/virgl_hw.h.97191BCB0E2552C8.idx differ
Binary files a/.cache/clangd/index/virgl_protocol.h.3A58776C31E2FFA1.idx and b/.cache/clangd/index/virgl_protocol.h.3A58776C31E2FFA1.idx differ
Binary files a/.cache/clangd/index/virgl_resource.c.C3078CF893894205.idx and b/.cache/clangd/index/virgl_resource.c.C3078CF893894205.idx differ
Binary files a/.cache/clangd/index/virgl_resource.h.0F020483DD7A6021.idx and b/.cache/clangd/index/virgl_resource.h.0F020483DD7A6021.idx differ
Binary files a/.cache/clangd/index/virgl_util.c.4788350FFFDD771F.idx and b/.cache/clangd/index/virgl_util.c.4788350FFFDD771F.idx differ
Binary files a/.cache/clangd/index/virgl_util.h.8DE07BD4E4FF3873.idx and b/.cache/clangd/index/virgl_util.h.8DE07BD4E4FF3873.idx differ
Binary files a/.cache/clangd/index/virglrenderer_hw.h.1BD264020332BA87.idx and b/.cache/clangd/index/virglrenderer_hw.h.1BD264020332BA87.idx differ
Binary files a/.cache/clangd/index/virglrenderer.c.D66E26A338E7C96A.idx and b/.cache/clangd/index/virglrenderer.c.D66E26A338E7C96A.idx differ
Binary files a/.cache/clangd/index/virglrenderer.h.B5D233ABCC026A26.idx and b/.cache/clangd/index/virglrenderer.h.B5D233ABCC026A26.idx differ
Binary files a/.cache/clangd/index/vk_platform.h.71EEBFBAADF9E881.idx and b/.cache/clangd/index/vk_platform.h.71EEBFBAADF9E881.idx differ
Binary files a/.cache/clangd/index/vkr_allocator.c.496F287038C332C2.idx and b/.cache/clangd/index/vkr_allocator.c.496F287038C332C2.idx differ
Binary files a/.cache/clangd/index/vkr_allocator.h.2F7B3CDE9FDDF2BE.idx and b/.cache/clangd/index/vkr_allocator.h.2F7B3CDE9FDDF2BE.idx differ
Binary files a/.cache/clangd/index/vkr_buffer.c.C919583EAE771D2A.idx and b/.cache/clangd/index/vkr_buffer.c.C919583EAE771D2A.idx differ
Binary files a/.cache/clangd/index/vkr_buffer.h.000D8C93F0F4519A.idx and b/.cache/clangd/index/vkr_buffer.h.000D8C93F0F4519A.idx differ
Binary files a/.cache/clangd/index/vkr_command_buffer.c.426FDFE4AD4C246B.idx and b/.cache/clangd/index/vkr_command_buffer.c.426FDFE4AD4C246B.idx differ
Binary files a/.cache/clangd/index/vkr_command_buffer.h.1ABE4781F5C6BFEA.idx and b/.cache/clangd/index/vkr_command_buffer.h.1ABE4781F5C6BFEA.idx differ
Binary files a/.cache/clangd/index/vkr_common.c.5FAF85CFAC699E16.idx and b/.cache/clangd/index/vkr_common.c.5FAF85CFAC699E16.idx differ
Binary files a/.cache/clangd/index/vkr_common.h.2245B48DAEBA8DC6.idx and b/.cache/clangd/index/vkr_common.h.2245B48DAEBA8DC6.idx differ
Binary files a/.cache/clangd/index/vkr_context.c.19E46C075AFD6BCC.idx and b/.cache/clangd/index/vkr_context.c.19E46C075AFD6BCC.idx differ
Binary files a/.cache/clangd/index/vkr_context.h.9D839FFDE4F796AA.idx and b/.cache/clangd/index/vkr_context.h.9D839FFDE4F796AA.idx differ
Binary files a/.cache/clangd/index/vkr_cs.c.E39AB3B940BD81F4.idx and b/.cache/clangd/index/vkr_cs.c.E39AB3B940BD81F4.idx differ
Binary files a/.cache/clangd/index/vkr_cs.h.F878ACB5D8768701.idx and b/.cache/clangd/index/vkr_cs.h.F878ACB5D8768701.idx differ
Binary files a/.cache/clangd/index/vkr_descriptor_set.c.847ED4FF0458F4A2.idx and b/.cache/clangd/index/vkr_descriptor_set.c.847ED4FF0458F4A2.idx differ
Binary files a/.cache/clangd/index/vkr_descriptor_set.h.ABE5ED5B8CB7C5CD.idx and b/.cache/clangd/index/vkr_descriptor_set.h.ABE5ED5B8CB7C5CD.idx differ
Binary files a/.cache/clangd/index/vkr_device_memory.c.E534A52D66A93E13.idx and b/.cache/clangd/index/vkr_device_memory.c.E534A52D66A93E13.idx differ
Binary files a/.cache/clangd/index/vkr_device_memory.h.E93A5081C2BFABE1.idx and b/.cache/clangd/index/vkr_device_memory.h.E93A5081C2BFABE1.idx differ
Binary files a/.cache/clangd/index/vkr_device.c.FF03FF0C5289F258.idx and b/.cache/clangd/index/vkr_device.c.FF03FF0C5289F258.idx differ
Binary files a/.cache/clangd/index/vkr_device.h.6B2EC5163D328B93.idx and b/.cache/clangd/index/vkr_device.h.6B2EC5163D328B93.idx differ
Binary files a/.cache/clangd/index/vkr_image.c.7E2F42D7E1B19CEC.idx and b/.cache/clangd/index/vkr_image.c.7E2F42D7E1B19CEC.idx differ
Binary files a/.cache/clangd/index/vkr_image.h.8C6A732D7C23E14C.idx and b/.cache/clangd/index/vkr_image.h.8C6A732D7C23E14C.idx differ
Binary files a/.cache/clangd/index/vkr_instance.c.6480EA71F5BCE477.idx and b/.cache/clangd/index/vkr_instance.c.6480EA71F5BCE477.idx differ
Binary files a/.cache/clangd/index/vkr_instance.h.2E88F4D38F840BEE.idx and b/.cache/clangd/index/vkr_instance.h.2E88F4D38F840BEE.idx differ
Binary files a/.cache/clangd/index/vkr_physical_device.c.9BF195335350A5A5.idx and b/.cache/clangd/index/vkr_physical_device.c.9BF195335350A5A5.idx differ
Binary files a/.cache/clangd/index/vkr_physical_device.h.63146EBD661CC0F5.idx and b/.cache/clangd/index/vkr_physical_device.h.63146EBD661CC0F5.idx differ
Binary files a/.cache/clangd/index/vkr_pipeline.c.7D219885F1EE8853.idx and b/.cache/clangd/index/vkr_pipeline.c.7D219885F1EE8853.idx differ
Binary files a/.cache/clangd/index/vkr_pipeline.h.AB6B2EA82260291D.idx and b/.cache/clangd/index/vkr_pipeline.h.AB6B2EA82260291D.idx differ
Binary files a/.cache/clangd/index/vkr_query_pool.c.1E1EC80D3862E836.idx and b/.cache/clangd/index/vkr_query_pool.c.1E1EC80D3862E836.idx differ
Binary files a/.cache/clangd/index/vkr_query_pool.h.FF5AC8924ECECE6F.idx and b/.cache/clangd/index/vkr_query_pool.h.FF5AC8924ECECE6F.idx differ
Binary files a/.cache/clangd/index/vkr_queue.c.07CA1F018AED98A1.idx and b/.cache/clangd/index/vkr_queue.c.07CA1F018AED98A1.idx differ
Binary files a/.cache/clangd/index/vkr_queue.h.A36C1180A1D973AA.idx and b/.cache/clangd/index/vkr_queue.h.A36C1180A1D973AA.idx differ
Binary files a/.cache/clangd/index/vkr_render_pass.c.373CABD925C30BB3.idx and b/.cache/clangd/index/vkr_render_pass.c.373CABD925C30BB3.idx differ
Binary files a/.cache/clangd/index/vkr_render_pass.h.3BFD51339F9D42A6.idx and b/.cache/clangd/index/vkr_render_pass.h.3BFD51339F9D42A6.idx differ
Binary files a/.cache/clangd/index/vkr_renderer.c.D836B07B10D9D055.idx and b/.cache/clangd/index/vkr_renderer.c.D836B07B10D9D055.idx differ
Binary files a/.cache/clangd/index/vkr_renderer.h.DA36D1E2FBBAA0FF.idx and b/.cache/clangd/index/vkr_renderer.h.DA36D1E2FBBAA0FF.idx differ
Binary files a/.cache/clangd/index/vkr_ring.c.7E1576E857566908.idx and b/.cache/clangd/index/vkr_ring.c.7E1576E857566908.idx differ
Binary files a/.cache/clangd/index/vkr_ring.h.98EEB1A72D7305CD.idx and b/.cache/clangd/index/vkr_ring.h.98EEB1A72D7305CD.idx differ
Binary files a/.cache/clangd/index/vkr_transport.c.42DFAD061440672A.idx and b/.cache/clangd/index/vkr_transport.c.42DFAD061440672A.idx differ
Binary files a/.cache/clangd/index/vkr_transport.h.0D331218F9F5F750.idx and b/.cache/clangd/index/vkr_transport.h.0D331218F9F5F750.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_buffer_view.h.9891D079F59DE792.idx and b/.cache/clangd/index/vn_protocol_renderer_buffer_view.h.9891D079F59DE792.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_buffer.h.76E8BAF00355F0E3.idx and b/.cache/clangd/index/vn_protocol_renderer_buffer.h.76E8BAF00355F0E3.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_command_buffer.h.5F1C580CC116763D.idx and b/.cache/clangd/index/vn_protocol_renderer_command_buffer.h.5F1C580CC116763D.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_command_pool.h.EDC9616D5DF1FF83.idx and b/.cache/clangd/index/vn_protocol_renderer_command_pool.h.EDC9616D5DF1FF83.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_cs.h.CC1AEF97F1684873.idx and b/.cache/clangd/index/vn_protocol_renderer_cs.h.CC1AEF97F1684873.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_defines.h.F6E05DABAD912DE2.idx and b/.cache/clangd/index/vn_protocol_renderer_defines.h.F6E05DABAD912DE2.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_descriptor_pool.h.3B871932FBD13958.idx and b/.cache/clangd/index/vn_protocol_renderer_descriptor_pool.h.3B871932FBD13958.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_descriptor_set_layout.h.2DF51545B08053B1.idx and b/.cache/clangd/index/vn_protocol_renderer_descriptor_set_layout.h.2DF51545B08053B1.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_descriptor_set.h.62FE1310D0391D01.idx and b/.cache/clangd/index/vn_protocol_renderer_descriptor_set.h.62FE1310D0391D01.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_descriptor_update_template.h.ABFE634DF1A5D9B1.idx and b/.cache/clangd/index/vn_protocol_renderer_descriptor_update_template.h.ABFE634DF1A5D9B1.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_device_memory.h.2279842A359EF536.idx and b/.cache/clangd/index/vn_protocol_renderer_device_memory.h.2279842A359EF536.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_device.h.CF2F14F03A8FE312.idx and b/.cache/clangd/index/vn_protocol_renderer_device.h.CF2F14F03A8FE312.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_dispatches.h.CC3FD3B8DB81604D.idx and b/.cache/clangd/index/vn_protocol_renderer_dispatches.h.CC3FD3B8DB81604D.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_event.h.D55D5AC73DA4DAA0.idx and b/.cache/clangd/index/vn_protocol_renderer_event.h.D55D5AC73DA4DAA0.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_fence.h.37D7259790500FED.idx and b/.cache/clangd/index/vn_protocol_renderer_fence.h.37D7259790500FED.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_framebuffer.h.1275BA46E87DF020.idx and b/.cache/clangd/index/vn_protocol_renderer_framebuffer.h.1275BA46E87DF020.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_handles.h.54CFA91448EF6A3E.idx and b/.cache/clangd/index/vn_protocol_renderer_handles.h.54CFA91448EF6A3E.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_image_view.h.FBE14E19315E5AED.idx and b/.cache/clangd/index/vn_protocol_renderer_image_view.h.FBE14E19315E5AED.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_image.h.4E546BC244586A2D.idx and b/.cache/clangd/index/vn_protocol_renderer_image.h.4E546BC244586A2D.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_info.h.CC3594FBE8EE624D.idx and b/.cache/clangd/index/vn_protocol_renderer_info.h.CC3594FBE8EE624D.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_instance.h.CB61FA980A89FDB0.idx and b/.cache/clangd/index/vn_protocol_renderer_instance.h.CB61FA980A89FDB0.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_pipeline_cache.h.F6971C915E12120E.idx and b/.cache/clangd/index/vn_protocol_renderer_pipeline_cache.h.F6971C915E12120E.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_pipeline_layout.h.D5BBEB2AD4399D60.idx and b/.cache/clangd/index/vn_protocol_renderer_pipeline_layout.h.D5BBEB2AD4399D60.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_pipeline.h.C2B36E2911B017ED.idx and b/.cache/clangd/index/vn_protocol_renderer_pipeline.h.C2B36E2911B017ED.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_private_data_slot.h.9A4C436C563AA2D3.idx and b/.cache/clangd/index/vn_protocol_renderer_private_data_slot.h.9A4C436C563AA2D3.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_query_pool.h.72EDC9D2C602F43E.idx and b/.cache/clangd/index/vn_protocol_renderer_query_pool.h.72EDC9D2C602F43E.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_queue.h.E01F254F6E6056CC.idx and b/.cache/clangd/index/vn_protocol_renderer_queue.h.E01F254F6E6056CC.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_render_pass.h.06A56AD44A6527CD.idx and b/.cache/clangd/index/vn_protocol_renderer_render_pass.h.06A56AD44A6527CD.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_sampler_ycbcr_conversion.h.CAAE60EE76E98736.idx and b/.cache/clangd/index/vn_protocol_renderer_sampler_ycbcr_conversion.h.CAAE60EE76E98736.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_sampler.h.5232FBA69E77FC6F.idx and b/.cache/clangd/index/vn_protocol_renderer_sampler.h.5232FBA69E77FC6F.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_semaphore.h.B653725110A4FC9D.idx and b/.cache/clangd/index/vn_protocol_renderer_semaphore.h.B653725110A4FC9D.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_shader_module.h.037A15CE52DD17B5.idx and b/.cache/clangd/index/vn_protocol_renderer_shader_module.h.037A15CE52DD17B5.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_structs.h.3BA2406BB157C02D.idx and b/.cache/clangd/index/vn_protocol_renderer_structs.h.3BA2406BB157C02D.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_transport.h.081766C75D3012EF.idx and b/.cache/clangd/index/vn_protocol_renderer_transport.h.081766C75D3012EF.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_types.h.D2C7C6D5A1C85568.idx and b/.cache/clangd/index/vn_protocol_renderer_types.h.D2C7C6D5A1C85568.idx differ
Binary files a/.cache/clangd/index/vn_protocol_renderer_util.h.E6F0253214813B2C.idx and b/.cache/clangd/index/vn_protocol_renderer_util.h.E6F0253214813B2C.idx differ
Binary files a/.cache/clangd/index/vrend_blitter.c.9B1355FBF60EF71B.idx and b/.cache/clangd/index/vrend_blitter.c.9B1355FBF60EF71B.idx differ
Binary files a/.cache/clangd/index/vrend_blitter.h.0BABD1EA028144EC.idx and b/.cache/clangd/index/vrend_blitter.h.0BABD1EA028144EC.idx differ
Binary files a/.cache/clangd/index/vrend_debug.c.23371E2A548AE02E.idx and b/.cache/clangd/index/vrend_debug.c.23371E2A548AE02E.idx differ
Binary files a/.cache/clangd/index/vrend_debug.h.59696ABBEBAC9900.idx and b/.cache/clangd/index/vrend_debug.h.59696ABBEBAC9900.idx differ
Binary files a/.cache/clangd/index/vrend_decode.c.070CE8C49BF4DF54.idx and b/.cache/clangd/index/vrend_decode.c.070CE8C49BF4DF54.idx differ
Binary files a/.cache/clangd/index/vrend_formats.c.01596907CCBA3322.idx and b/.cache/clangd/index/vrend_formats.c.01596907CCBA3322.idx differ
Binary files a/.cache/clangd/index/vrend_iov.h.1555D1B0256F2F02.idx and b/.cache/clangd/index/vrend_iov.h.1555D1B0256F2F02.idx differ
Binary files a/.cache/clangd/index/vrend_object.c.12345B1E3C2B8E6E.idx and b/.cache/clangd/index/vrend_object.c.12345B1E3C2B8E6E.idx differ
Binary files a/.cache/clangd/index/vrend_object.h.1F662D56BAFBEBF0.idx and b/.cache/clangd/index/vrend_object.h.1F662D56BAFBEBF0.idx differ
Binary files a/.cache/clangd/index/vrend_renderer.c.650B3DD6A5561CC2.idx and b/.cache/clangd/index/vrend_renderer.c.650B3DD6A5561CC2.idx differ
Binary files a/.cache/clangd/index/vrend_renderer.h.E751C5EFE3762D2B.idx and b/.cache/clangd/index/vrend_renderer.h.E751C5EFE3762D2B.idx differ
Binary files a/.cache/clangd/index/vrend_shader.c.787F04F38A3643A3.idx and b/.cache/clangd/index/vrend_shader.c.787F04F38A3643A3.idx differ
Binary files a/.cache/clangd/index/vrend_shader.h.25438B5CCEA3177E.idx and b/.cache/clangd/index/vrend_shader.h.25438B5CCEA3177E.idx differ
Binary files a/.cache/clangd/index/vrend_strbuf.h.DED14CE905ECB22A.idx and b/.cache/clangd/index/vrend_strbuf.h.DED14CE905ECB22A.idx differ
Binary files a/.cache/clangd/index/vrend_tweaks.c.F0506809929C75EC.idx and b/.cache/clangd/index/vrend_tweaks.c.F0506809929C75EC.idx differ
Binary files a/.cache/clangd/index/vrend_tweaks.h.6D0312C947FE3695.idx and b/.cache/clangd/index/vrend_tweaks.h.6D0312C947FE3695.idx differ
Binary files a/.cache/clangd/index/vrend_winsys.c.15A8F37409DCB609.idx and b/.cache/clangd/index/vrend_winsys.c.15A8F37409DCB609.idx differ
Binary files a/.cache/clangd/index/vrend_winsys.h.EA2D80A7C064EC3B.idx and b/.cache/clangd/index/vrend_winsys.h.EA2D80A7C064EC3B.idx differ
Binary files a/.cache/clangd/index/vtest_protocol.h.77D5C3B217758B0D.idx and b/.cache/clangd/index/vtest_protocol.h.77D5C3B217758B0D.idx differ
Binary files a/.cache/clangd/index/vtest_renderer.c.81879BD960BF7598.idx and b/.cache/clangd/index/vtest_renderer.c.81879BD960BF7598.idx differ
Binary files a/.cache/clangd/index/vtest_server.c.7097FDCC6FC5700B.idx and b/.cache/clangd/index/vtest_server.c.7097FDCC6FC5700B.idx differ
Binary files a/.cache/clangd/index/vtest_shm.c.9DD43A85FA8DD34A.idx and b/.cache/clangd/index/vtest_shm.c.9DD43A85FA8DD34A.idx differ
Binary files a/.cache/clangd/index/vtest_shm.h.A5D9F6B5A9FCD2D5.idx and b/.cache/clangd/index/vtest_shm.h.A5D9F6B5A9FCD2D5.idx differ
Binary files a/.cache/clangd/index/vtest.h.068E055978F85BBC.idx and b/.cache/clangd/index/vtest.h.068E055978F85BBC.idx differ
Binary files a/.cache/clangd/index/vulkan_core.h.2697CD94A89619F1.idx and b/.cache/clangd/index/vulkan_core.h.2697CD94A89619F1.idx differ
Binary files a/.cache/clangd/index/vulkan.h.91BC23EF19A18127.idx and b/.cache/clangd/index/vulkan.h.91BC23EF19A18127.idx differ
Binary files a/.cache/clangd/index/xxhash.h.855524FCC5C784D7.idx and b/.cache/clangd/index/xxhash.h.855524FCC5C784D7.idx differ
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/venus-protocol/vn_protocol_renderer_dispatches.h b/src/venus/venus-protocol/vn_protocol_renderer_dispatches.h
--- a/src/venus/venus-protocol/vn_protocol_renderer_dispatches.h
+++ b/src/venus/venus-protocol/vn_protocol_renderer_dispatches.h
@@ -555,6 +555,7 @@
     vn_decode_VkCommandTypeEXT(ctx->decoder, &cmd_type);
     vn_decode_VkFlags(ctx->decoder, &cmd_flags);
 
+
     {
 #ifdef DEBUG
         TRACE_SCOPE_SLOW(vn_dispatch_command_name(cmd_type));
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_common.h b/src/venus/vkr_common.h
--- a/src/venus/vkr_common.h
+++ b/src/venus/vkr_common.h
@@ -54,6 +54,50 @@
       return obj;                                                                        \
    }
 
+#ifdef __APPLE__
+/* VK_EXT_host_image_copy, declared here because the vendored headers predate it.
+ *
+ * This is what makes the guest's pixels readable at all: the host renders into
+ * an optimal-tiled image, which on this platform lives in private texture
+ * memory, so no mapping of the underlying allocation ever shows the picture.
+ * This extension reads such an image straight into a linear host buffer.
+ */
+#define VKR_IMAGE_USAGE_HOST_TRANSFER_BIT 0x00400000u
+#define VKR_STRUCTURE_TYPE_PHYSICAL_DEVICE_HOST_IMAGE_COPY_FEATURES 1000270000
+#define VKR_STRUCTURE_TYPE_IMAGE_TO_MEMORY_COPY 1000270003
+#define VKR_STRUCTURE_TYPE_COPY_IMAGE_TO_MEMORY_INFO 1000270004
+
+typedef struct VkrPhysicalDeviceHostImageCopyFeatures {
+   VkStructureType sType;
+   void *pNext;
+   VkBool32 hostImageCopy;
+} VkrPhysicalDeviceHostImageCopyFeatures;
+
+typedef struct VkrImageToMemoryCopy {
+   VkStructureType sType;
+   const void *pNext;
+   void *pHostPointer;
+   uint32_t memoryRowLength;
+   uint32_t memoryImageHeight;
+   VkImageSubresourceLayers imageSubresource;
+   VkOffset3D imageOffset;
+   VkExtent3D imageExtent;
+} VkrImageToMemoryCopy;
+
+typedef struct VkrCopyImageToMemoryInfo {
+   VkStructureType sType;
+   const void *pNext;
+   VkFlags flags;
+   VkImage srcImage;
+   VkImageLayout srcImageLayout;
+   uint32_t regionCount;
+   const VkrImageToMemoryCopy *pRegions;
+} VkrCopyImageToMemoryInfo;
+
+typedef VkResult(VKAPI_PTR *PFN_vkrCopyImageToMemory)(VkDevice,
+                                                      const VkrCopyImageToMemoryInfo *);
+#endif /* __APPLE__ */
+
 /* vkr_region_is_valid should be used to check for overflows */
 #define VKR_REGION_INIT(offset, size)                                                    \
    {                                                                                     \
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_context.c b/src/venus/vkr_context.c
--- a/src/venus/vkr_context.c
+++ b/src/venus/vkr_context.c
@@ -324,7 +324,14 @@
     * - vkGetMemoryFdPropertiesKHR for dma_buf fd properties query
     */
    int res_fd = -1;
+#ifdef __APPLE__
+   /* A cross-device blob is shared through the mapped host pointer here, not an
+    * fd, so there is nothing to duplicate.
+    */
+   if (false) {
+#else
    if (blob_flags & VIRGL_RENDERER_BLOB_FLAG_USE_CROSS_DEVICE) {
+#endif
       res_fd = os_dupfd_cloexec(blob.u.fd);
       if (res_fd < 0) {
          close(blob.u.fd);
@@ -332,14 +339,29 @@
       }
    }
 
+#ifdef __APPLE__
+   void *const res_map_ptr = (void *)(uintptr_t)blob.map_ptr;
+#else
+   void *const res_map_ptr = NULL;
+#endif
+
    if (!vkr_context_import_resource_internal(ctx, res_id, blob_size, blob.type, res_fd,
-                                             NULL)) {
+                                             res_map_ptr)) {
       if (res_fd >= 0)
          close(res_fd);
       close(blob.u.fd);
       return false;
    }
 
+#ifdef __APPLE__
+   {
+      struct vkr_resource *res = vkr_context_get_resource(ctx, res_id);
+      if (res)
+         res->mem = mem;
+         res->mem_id = mem->base.id;
+   }
+#endif
+
    *out_blob = blob;
 
    return true;
@@ -605,6 +627,7 @@
 
    _mesa_hash_table_destroy(ctx->object_table, vkr_context_free_object);
    mtx_destroy(&ctx->object_mutex);
+   mtx_destroy(&ctx->scanout_mutex);
 
    vkr_cs_encoder_fini(&ctx->encoder);
    vkr_cs_decoder_fini(&ctx->decoder);
@@ -664,6 +687,8 @@
    if (!vkr_context_wait_ring_init(ctx))
       goto err_ctx_wait_ring_init;
 
+   if (mtx_init(&ctx->scanout_mutex, mtx_plain) != thrd_success)
+      goto err_scanout_mutex;
    if (mtx_init(&ctx->object_mutex, mtx_plain) != thrd_success)
       goto err_ctx_object_mutex;
 
@@ -703,6 +728,8 @@
 err_ctx_object_table:
    mtx_destroy(&ctx->object_mutex);
 err_ctx_object_mutex:
+   mtx_destroy(&ctx->scanout_mutex);
+err_scanout_mutex:
    vkr_context_wait_ring_fini(ctx);
 err_ctx_wait_ring_init:
    free(ctx->debug_name);
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_context.h b/src/venus/vkr_context.h
--- a/src/venus/vkr_context.h
+++ b/src/venus/vkr_context.h
@@ -20,6 +20,15 @@
 struct vkr_resource {
    uint32_t res_id;
 
+#ifdef __APPLE__
+   /* the allocation this resource was exported from, so a scanout read can
+    * find the image bound to it */
+   struct vkr_device_memory *mem;
+   /* its object id: the memory can die with its device without this
+    * resource hearing about it, so a reader checks the id still resolves */
+   vkr_object_id mem_id;
+#endif
+
    enum virgl_resource_fd_type fd_type;
 
    union {
@@ -78,6 +87,8 @@
    } ring_monitor;
 
    mtx_t object_mutex;
+   /* held across a scanout readback and across image/memory destruction */
+   mtx_t scanout_mutex;
    struct hash_table *object_table;
 
    mtx_t resource_mutex;
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_device_memory.c b/src/venus/vkr_device_memory.c
--- a/src/venus/vkr_device_memory.c
+++ b/src/venus/vkr_device_memory.c
@@ -3,7 +3,10 @@
  * SPDX-License-Identifier: MIT
  */
 
+#include <stdio.h>
+
 #include "vkr_device_memory.h"
+#include "vkr_image.h"
 
 #include "venus-protocol/vn_protocol_renderer_transport.h"
 
@@ -149,17 +152,103 @@
    struct vkr_physical_device *physical_dev = dev->physical_device;
 
    VkMemoryAllocateInfo *alloc_info = (VkMemoryAllocateInfo *)args->pAllocateInfo;
-   const uint32_t mem_type_index = alloc_info->memoryTypeIndex;
+   uint32_t mem_type_index = alloc_info->memoryTypeIndex;
    if (unlikely(mem_type_index >= physical_dev->memory_properties.memoryTypeCount)) {
       args->ret = VK_ERROR_UNKNOWN;
       return;
    }
 
+#ifdef __APPLE__
+   /* Sharing a resource with the guest means handing over a mapped host
+    * pointer, because there are no fds here. Memory the host reports as
+    * device-local only cannot be mapped at all, so an allocation that lands
+    * there can never be shared -- which is what stopped a compositor from
+    * getting an output buffer.
+    *
+    * The host's memory is unified and every type sits in the same heap, so
+    * moving the allocation to a host-visible type costs nothing.
+    */
+   {
+      const VkPhysicalDeviceMemoryProperties *props = &physical_dev->memory_properties;
+      if (!(props->memoryTypes[mem_type_index].propertyFlags &
+            VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT)) {
+         for (uint32_t i = 0; i < props->memoryTypeCount; i++) {
+            const VkMemoryPropertyFlags f = props->memoryTypes[i].propertyFlags;
+            if ((f & VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT) &&
+                props->memoryTypes[i].heapIndex ==
+                   props->memoryTypes[mem_type_index].heapIndex) {
+               mem_type_index = i;
+               alloc_info->memoryTypeIndex = i;
+               break;
+            }
+         }
+      }
+   }
+#endif
+
+#ifdef __APPLE__
+   /* The host has no fd-based external memory, so an allocation that asks to
+    * export or import one fails and the memory object is never created -- and
+    * the bind that follows then references a handle that does not exist, which
+    * kills the command stream. The sharing actually happens through the
+    * virtio-gpu resource, so drop the request the host cannot honour.
+    */
+   {
+      const VkStructureType drop[] = {
+         VK_STRUCTURE_TYPE_EXPORT_MEMORY_ALLOCATE_INFO,
+         VK_STRUCTURE_TYPE_IMPORT_MEMORY_FD_INFO_KHR,
+      };
+      for (uint32_t i = 0; i < ARRAY_SIZE(drop); i++) {
+         VkBaseInStructure *prev = vkr_find_prev_struct(alloc_info, drop[i]);
+         if (prev)
+            prev->pNext = prev->pNext->pNext;
+      }
+   }
+#endif
+
    /* translate VkImportMemoryResourceInfoMESA into VkImportMemoryFdInfoKHR in place */
    VkImportMemoryFdInfoKHR local_import_info = { .fd = -1 };
    VkImportMemoryResourceInfoMESA *res_info = NULL;
    VkBaseInStructure *prev_of_res_info = vkr_find_prev_struct(
       alloc_info, VK_STRUCTURE_TYPE_IMPORT_MEMORY_RESOURCE_INFO_MESA);
+
+#ifdef __APPLE__
+   /* The guest is asking for memory that aliases a resource it already owns.
+    * There is no fd to import here, so wrap the pointer the resource is mapped
+    * at instead -- the host allocation then refers to the very same storage.
+    */
+   VkImportMemoryHostPointerInfoEXT local_host_ptr_info;
+   if (prev_of_res_info && physical_dev->EXT_external_memory_host) {
+      const VkImportMemoryResourceInfoMESA *info =
+         (const VkImportMemoryResourceInfoMESA *)prev_of_res_info->pNext;
+      const struct vkr_resource *res = vkr_context_get_resource(ctx, info->resourceId);
+      if (res && res->u.data) {
+         local_host_ptr_info = (VkImportMemoryHostPointerInfoEXT){
+            .sType = VK_STRUCTURE_TYPE_IMPORT_MEMORY_HOST_POINTER_INFO_EXT,
+            .pNext = info->pNext,
+            .handleType = VK_EXTERNAL_MEMORY_HANDLE_TYPE_HOST_ALLOCATION_BIT_EXT,
+            .pHostPointer = res->u.data,
+         };
+         prev_of_res_info->pNext = (const struct VkBaseInStructure *)&local_host_ptr_info;
+
+         struct vkr_device_memory *host_mem = vkr_device_memory_create_and_add(ctx, args);
+         if (!host_mem)
+            return;
+
+         host_mem->device = dev;
+         list_inithead(&host_mem->bound_images);
+         host_mem->guest_ptr = res->u.data;
+         host_mem->property_flags =
+            physical_dev->memory_properties.memoryTypes[mem_type_index].propertyFlags;
+         host_mem->valid_fd_types = 0;
+         host_mem->gbm_bo = NULL;
+         host_mem->allocation_size = alloc_info->allocationSize;
+         host_mem->memory_type_index = mem_type_index;
+         return;
+      }
+   }
+#endif
+
    if (prev_of_res_info) {
       res_info = (VkImportMemoryResourceInfoMESA *)prev_of_res_info->pNext;
       if (!vkr_get_fd_info_from_resource_info(ctx, res_info, &local_import_info)) {
@@ -252,6 +341,7 @@
    }
 
    mem->device = dev;
+   list_inithead(&mem->bound_images);
    mem->property_flags = property_flags;
    mem->valid_fd_types = valid_fd_types;
    mem->gbm_bo = gbm_bo;
@@ -263,6 +353,18 @@
 vkr_dispatch_vkFreeMemory(struct vn_dispatch_context *dispatch,
                           struct vn_command_vkFreeMemory *args)
 {
+   struct vkr_context *scanout_ctx = dispatch->data;
+   mtx_lock(&scanout_ctx->scanout_mutex);
+   {
+      struct vkr_device_memory *freeing = vkr_device_memory_from_handle(args->memory);
+      if (freeing && freeing->bound_images.next) {
+         list_for_each_entry_safe (struct vkr_image, bimg, &freeing->bound_images, mem_head) {
+            list_del(&bimg->mem_head);
+            bimg->bound_mem = NULL;
+         }
+         freeing->bound_image = NULL;
+      }
+   }
    struct vkr_device_memory *mem = vkr_device_memory_from_handle(args->memory);
    if (!mem)
       return;
@@ -275,6 +377,8 @@
 
    vkr_device_memory_release(mem);
    vkr_device_memory_destroy_and_remove(dispatch->data, args);
+out:
+   mtx_unlock(&scanout_ctx->scanout_mutex);
 }
 
 static void
@@ -382,6 +486,10 @@
     * to the same storage.
     */
    if (mem->exported) {
+#ifdef VKR_TRACE_COMMANDS
+      { FILE *tf = fopen("/tmp/vkr-res.log", "a");
+        if (tf) { fprintf(tf, "export_blob: ALREADY EXPORTED flags=0x%x\n", blob_flags); fclose(tf); } }
+#endif
       vkr_log("mem has been exported");
       return false;
    }
@@ -406,11 +514,7 @@
    enum virgl_resource_fd_type fd_type;
    VkExternalMemoryHandleTypeFlagBits handle_type;
    struct virgl_resource_vulkan_info vulkan_info;
-   if (blob_flags & VIRGL_RENDERER_BLOB_FLAG_USE_CROSS_DEVICE) {
-      if (!can_export_dma_buf) {
-         vkr_log("mem cannot export to dma_buf for cross device blob sharing");
-         return false;
-      }
+   if ((blob_flags & VIRGL_RENDERER_BLOB_FLAG_USE_CROSS_DEVICE) && can_export_dma_buf) {
       fd_type = VIRGL_RESOURCE_FD_DMABUF;
       handle_type = VK_EXTERNAL_MEMORY_HANDLE_TYPE_DMA_BUF_BIT_EXT;
    } else if (can_export_dma_buf) {
@@ -435,8 +539,17 @@
    } else {
 #ifdef __APPLE__
       void *ptr;
-      if (vkMapMemory(mem->device->base.handle.device, mem->base.handle.device_memory,
-                      0, mem->allocation_size, 0, &ptr) != VK_SUCCESS) {
+      VkResult map_res = vkMapMemory(mem->device->base.handle.device,
+                                     mem->base.handle.device_memory, 0,
+                                     mem->allocation_size, 0, &ptr);
+#ifdef VKR_TRACE_COMMANDS
+      {
+         FILE *tf = fopen("/tmp/vkr-res.log", "a");
+         if (tf) { fprintf(tf, "export_blob: flags=0x%x props=0x%x map=%d\n", blob_flags,
+                           mem->property_flags, (int)map_res); fclose(tf); }
+      }
+#endif
+      if (map_res != VK_SUCCESS) {
          vkr_log("vkMapMemory failed");
          return false;
       }
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_device_memory.h b/src/venus/vkr_device_memory.h
--- a/src/venus/vkr_device_memory.h
+++ b/src/venus/vkr_device_memory.h
@@ -13,6 +13,16 @@
 struct vkr_device_memory {
    struct vkr_object base;
 
+#ifdef __APPLE__
+   /* the image bound here most recently, and every image bound here --
+    * see vkr_dispatch_vkBindImageMemory2 */
+   struct vkr_image *bound_image;
+   struct list_head bound_images;
+   /* The guest storage this memory wraps, so a scanout read can tell the
+    * desktop's own buffers apart from an app's same-size buffers. */
+   const void *guest_ptr;
+#endif
+
    struct vkr_device *device;
    uint32_t property_flags;
    uint32_t valid_fd_types;
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_device.c b/src/venus/vkr_device.c
--- a/src/venus/vkr_device.c
+++ b/src/venus/vkr_device.c
@@ -84,12 +84,34 @@
    struct vkr_physical_device *physical_dev =
       vkr_physical_device_from_handle(args->physicalDevice);
 
+#ifdef __APPLE__
+   /* Drop the extensions we advertise but the host does not implement. Left in
+    * place they reach MoltenVK, which fails the whole device create with
+    * VK_ERROR_EXTENSION_NOT_PRESENT.
+    */
+   {
+      VkDeviceCreateInfo *info = (VkDeviceCreateInfo *)args->pCreateInfo;
+      const char **names = (const char **)info->ppEnabledExtensionNames;
+      uint32_t kept = 0;
+      for (uint32_t i = 0; i < info->enabledExtensionCount; i++) {
+         if (vkr_physical_device_extension_is_emulated(names[i]))
+            continue;
+         names[kept++] = names[i];
+      }
+      info->enabledExtensionCount = kept;
+   }
+#endif
+
    /* append extensions for our own use */
    const char **exts = NULL;
    uint32_t ext_count = args->pCreateInfo->enabledExtensionCount;
    ext_count += physical_dev->KHR_external_memory_fd;
    ext_count += physical_dev->EXT_external_memory_dma_buf;
    ext_count += physical_dev->KHR_external_fence_fd;
+#ifdef __APPLE__
+   ext_count += physical_dev->EXT_external_memory_host;
+   ext_count += physical_dev->EXT_host_image_copy;
+#endif
    if (ext_count > args->pCreateInfo->enabledExtensionCount) {
       exts = malloc(sizeof(*exts) * ext_count);
       if (!exts) {
@@ -106,6 +128,12 @@
          exts[ext_count++] = "VK_EXT_external_memory_dma_buf";
       if (physical_dev->KHR_external_fence_fd)
          exts[ext_count++] = "VK_KHR_external_fence_fd";
+#ifdef __APPLE__
+      if (physical_dev->EXT_external_memory_host)
+         exts[ext_count++] = "VK_EXT_external_memory_host";
+      if (physical_dev->EXT_host_image_copy)
+         exts[ext_count++] = "VK_EXT_host_image_copy";
+#endif
 
       ((VkDeviceCreateInfo *)args->pCreateInfo)->ppEnabledExtensionNames = exts;
       ((VkDeviceCreateInfo *)args->pCreateInfo)->enabledExtensionCount = ext_count;
@@ -119,9 +147,36 @@
       return;
    }
 
+#ifdef __APPLE__
+   /* The host would fail the device create outright if asked for a feature it
+    * reports as unsupported. See vkr_dispatch_vkGetPhysicalDeviceFeatures2.
+    */
+   {
+      VkPhysicalDeviceRobustness2FeaturesEXT *rb2 = vkr_find_struct(
+         args->pCreateInfo->pNext,
+         VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ROBUSTNESS_2_FEATURES_EXT);
+      if (rb2)
+         rb2->nullDescriptor = VK_FALSE;
+   }
+#endif
+
+#ifdef __APPLE__
+   VkrPhysicalDeviceHostImageCopyFeatures host_copy_feature = {
+      .sType = VKR_STRUCTURE_TYPE_PHYSICAL_DEVICE_HOST_IMAGE_COPY_FEATURES,
+      .hostImageCopy = VK_TRUE,
+   };
+   if (physical_dev->EXT_host_image_copy) {
+      VkDeviceCreateInfo *info = (VkDeviceCreateInfo *)args->pCreateInfo;
+      host_copy_feature.pNext = (void *)info->pNext;
+      info->pNext = &host_copy_feature;
+   }
+#endif
+
    vn_replace_vkCreateDevice_args_handle(args);
    args->ret = vkCreateDevice(args->physicalDevice, args->pCreateInfo, NULL,
                               &dev->base.handle.device);
+#ifdef __APPLE__
+#endif
    if (args->ret != VK_SUCCESS) {
       free(exts);
       free(dev);
@@ -295,7 +350,11 @@
    if (!dev)
       return;
 
+   /* a device teardown frees every image and memory at once; no scanout
+    * read may be copying out of one meanwhile */
+   mtx_lock(&ctx->scanout_mutex);
    vkr_device_destroy(ctx, dev);
+   mtx_unlock(&ctx->scanout_mutex);
 }
 
 static void
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_image.c b/src/venus/vkr_image.c
--- a/src/venus/vkr_image.c
+++ b/src/venus/vkr_image.c
@@ -5,9 +5,51 @@
 
 #include "vkr_image.h"
 
+#include "vkr_device_memory.h"
 #include "vkr_image_gen.h"
 #include "vkr_physical_device.h"
 
+#ifdef __APPLE__
+/* Enough of the common colour formats to describe a compositor's buffers. */
+static uint32_t
+vkr_format_bytes_per_pixel(VkFormat format)
+{
+   switch (format) {
+   case VK_FORMAT_R8G8B8A8_UNORM:
+   case VK_FORMAT_R8G8B8A8_SRGB:
+   case VK_FORMAT_B8G8R8A8_UNORM:
+   case VK_FORMAT_B8G8R8A8_SRGB:
+   case VK_FORMAT_A8B8G8R8_UNORM_PACK32:
+   case VK_FORMAT_A8B8G8R8_SRGB_PACK32:
+   case VK_FORMAT_A2R10G10B10_UNORM_PACK32:
+   case VK_FORMAT_A2B10G10R10_UNORM_PACK32:
+      return 4;
+   case VK_FORMAT_R8G8B8_UNORM:
+   case VK_FORMAT_B8G8R8_UNORM:
+      return 3;
+   case VK_FORMAT_R8G8_UNORM:
+   case VK_FORMAT_R5G6B5_UNORM_PACK16:
+   case VK_FORMAT_B5G6R5_UNORM_PACK16:
+      return 2;
+   case VK_FORMAT_R8_UNORM:
+      return 1;
+   case VK_FORMAT_R16G16B16A16_SFLOAT:
+   case VK_FORMAT_R16G16B16A16_UNORM:
+      return 8;
+   default:
+      return 0;
+   }
+}
+
+static uint64_t
+vkr_align_row_pitch(uint64_t pitch)
+{
+   /* 256 is the alignment scanout hardware and every DRM driver assume */
+   const uint64_t align = 256;
+   return (pitch + align - 1) & ~(align - 1);
+}
+#endif
+
 static void
 vkr_dispatch_vkCreateImage(struct vn_dispatch_context *dispatch,
                            struct vn_command_vkCreateImage *args)
@@ -31,14 +73,79 @@
     * situation because the app does not consider the memory external.
     */
 
-   vkr_image_create_and_add(dispatch->data, args);
+#ifdef __APPLE__
+   bool was_drm_modifier = false;
+   /* The host knows neither DRM modifiers nor dma_buf handles and fails the
+    * whole image create if either is left in the chain. Linear is the only
+    * modifier advertised, so ask the host for a linear image instead and drop
+    * the structures it cannot read. The guest reaches the pixels through the
+    * virtio-gpu resource, so only the resource identity has to agree, not the
+    * host's private layout.
+    */
+   {
+      VkImageCreateInfo *info = (VkImageCreateInfo *)args->pCreateInfo;
+
+      if (info->tiling == VK_IMAGE_TILING_DRM_FORMAT_MODIFIER_EXT) {
+         info->tiling = VK_IMAGE_TILING_OPTIMAL;
+         was_drm_modifier = true;
+
+         /* so the picture can be read back out of the private texture memory
+          * this image really lives in */
+         struct vkr_device *dev_for_usage = vkr_device_from_handle(args->device);
+         if (dev_for_usage->physical_device->EXT_host_image_copy)
+            info->usage |= VKR_IMAGE_USAGE_HOST_TRANSFER_BIT;
+      }
+
+      const VkStructureType drop[] = {
+         VK_STRUCTURE_TYPE_IMAGE_DRM_FORMAT_MODIFIER_LIST_CREATE_INFO_EXT,
+         VK_STRUCTURE_TYPE_IMAGE_DRM_FORMAT_MODIFIER_EXPLICIT_CREATE_INFO_EXT,
+         VK_STRUCTURE_TYPE_EXTERNAL_MEMORY_IMAGE_CREATE_INFO,
+      };
+      for (uint32_t i = 0; i < ARRAY_SIZE(drop); i++) {
+         VkBaseInStructure *prev = vkr_find_prev_struct(info, drop[i]);
+         if (prev)
+            prev->pNext = prev->pNext->pNext;
+      }
+   }
+#endif
+
+#ifdef __APPLE__
+   const VkImageCreateInfo saved = *args->pCreateInfo;
+#endif
+
+   struct vkr_image *img = vkr_image_create_and_add(dispatch->data, args);
+
+#ifdef __APPLE__
+   if (img && was_drm_modifier) {
+      img->drm_modifier = true;
+      img->width = saved.extent.width;
+      img->height = saved.extent.height;
+      img->bytes_per_pixel = vkr_format_bytes_per_pixel(saved.format);
+   }
+#endif
 }
 
 static void
 vkr_dispatch_vkDestroyImage(struct vn_dispatch_context *dispatch,
                             struct vn_command_vkDestroyImage *args)
 {
+   struct vkr_context *ctx = dispatch->data;
+   mtx_lock(&ctx->scanout_mutex);
+#ifdef __APPLE__
+   {
+      struct vkr_image *dying = vkr_image_from_handle(args->image);
+      if (dying && dying->bound_mem) {
+         struct vkr_device_memory *mem = dying->bound_mem;
+         list_del(&dying->mem_head);
+         dying->bound_mem = NULL;
+         if (mem->bound_image == dying)
+            mem->bound_image = list_is_empty(&mem->bound_images) ? NULL
+               : list_last_entry(&mem->bound_images, struct vkr_image, mem_head);
+      }
+   }
+#endif
    vkr_image_destroy_and_remove(dispatch->data, args);
+   mtx_unlock(&ctx->scanout_mutex);
 }
 
 static void
@@ -93,27 +200,74 @@
                                          args->pSparseMemoryRequirements);
 }
 
+#ifdef __APPLE__
+/* Remember which image ends up in which allocation: a scanout read starts
+ * from the resource, and the resource only knows its memory. Recorded only
+ * after the bind really happened, under the scanout lock, so a read on the
+ * other thread never picks an image that is not backed yet.
+ */
+static void
+vkr_image_record_binding(struct vkr_context *ctx,
+                         struct vkr_image *img,
+                         struct vkr_device_memory *mem)
+{
+   if (!img || !mem)
+      return;
+   mtx_lock(&ctx->scanout_mutex);
+   if (img->bound_mem)
+      list_del(&img->mem_head);
+   img->bound_mem = mem;
+   list_addtail(&img->mem_head, &mem->bound_images);
+   mem->bound_image = img;
+   mtx_unlock(&ctx->scanout_mutex);
+}
+#endif
+
 static void
-vkr_dispatch_vkBindImageMemory(UNUSED struct vn_dispatch_context *dispatch,
+vkr_dispatch_vkBindImageMemory(struct vn_dispatch_context *dispatch,
                                struct vn_command_vkBindImageMemory *args)
 {
    struct vkr_device *dev = vkr_device_from_handle(args->device);
    struct vn_device_proc_table *vk = &dev->proc_table;
+#ifdef __APPLE__
+   struct vkr_image *img = vkr_image_from_handle(args->image);
+   struct vkr_device_memory *mem = vkr_device_memory_from_handle(args->memory);
+#endif
 
    vn_replace_vkBindImageMemory_args_handle(args);
    args->ret =
       vk->BindImageMemory(args->device, args->image, args->memory, args->memoryOffset);
+#ifdef __APPLE__
+   if (args->ret == VK_SUCCESS)
+      vkr_image_record_binding(dispatch->data, img, mem);
+#endif
 }
 
 static void
-vkr_dispatch_vkBindImageMemory2(UNUSED struct vn_dispatch_context *dispatch,
+vkr_dispatch_vkBindImageMemory2(struct vn_dispatch_context *dispatch,
                                 struct vn_command_vkBindImageMemory2 *args)
 {
    struct vkr_device *dev = vkr_device_from_handle(args->device);
    struct vn_device_proc_table *vk = &dev->proc_table;
+#ifdef __APPLE__
+   /* the bind infos carry host handles after vn_replace, so resolve first */
+   const uint32_t bind_count = args->bindInfoCount < 8 ? args->bindInfoCount : 8;
+   struct vkr_image *imgs[8];
+   struct vkr_device_memory *mems[8];
+   for (uint32_t i = 0; i < bind_count; i++) {
+      imgs[i] = vkr_image_from_handle(args->pBindInfos[i].image);
+      mems[i] = vkr_device_memory_from_handle(args->pBindInfos[i].memory);
+   }
+#endif
 
    vn_replace_vkBindImageMemory2_args_handle(args);
    args->ret = vk->BindImageMemory2(args->device, args->bindInfoCount, args->pBindInfos);
+#ifdef __APPLE__
+   if (args->ret == VK_SUCCESS) {
+      for (uint32_t i = 0; i < bind_count; i++)
+         vkr_image_record_binding(dispatch->data, imgs[i], mems[i]);
+   }
+#endif
 }
 
 static void
@@ -124,6 +278,29 @@
    struct vkr_device *dev = vkr_device_from_handle(args->device);
    struct vn_device_proc_table *vk = &dev->proc_table;
 
+#ifdef __APPLE__
+   /* Reading the layout of an optimal-tiled image is not something Vulkan
+    * allows, and the host answers with a stride the guest cannot use (it kept
+    * discarding the buffer and retrying). Describe the tightly packed layout
+    * the guest expects; the host's private tiling is never visible to it.
+    */
+   {
+      const struct vkr_image *img = vkr_image_from_handle(args->image);
+      if (img && img->drm_modifier && img->bytes_per_pixel) {
+         const uint64_t row = vkr_align_row_pitch(
+            (uint64_t)img->width * img->bytes_per_pixel);
+         *args->pLayout = (VkSubresourceLayout){
+            .offset = 0,
+            .size = row * img->height,
+            .rowPitch = row,
+            .arrayPitch = row * img->height,
+            .depthPitch = row * img->height,
+         };
+         return;
+      }
+   }
+#endif
+
    vn_replace_vkGetImageSubresourceLayout_args_handle(args);
    vk->GetImageSubresourceLayout(args->device, args->image, args->pSubresource,
                                  args->pLayout);
@@ -134,12 +311,23 @@
    UNUSED struct vn_dispatch_context *dispatch,
    struct vn_command_vkGetImageDrmFormatModifierPropertiesEXT *args)
 {
+#ifdef __APPLE__
+   /* The host does not implement this extension, so the proc pointer is NULL
+    * and calling it takes the VM down. Every modifier image here was created by
+    * vkr_dispatch_vkCreateImage, so the answer is always the one modifier we
+    * advertise.
+    */
+   args->pProperties->drmFormatModifier = 0ull; /* VKR_DRM_FORMAT_MOD_HOST_OPAQUE */
+   args->ret = VK_SUCCESS;
+   return;
+#else
    struct vkr_device *dev = vkr_device_from_handle(args->device);
    struct vn_device_proc_table *vk = &dev->proc_table;
 
    vn_replace_vkGetImageDrmFormatModifierPropertiesEXT_args_handle(args);
    args->ret = vk->GetImageDrmFormatModifierPropertiesEXT(args->device, args->image,
                                                           args->pProperties);
+#endif
 }
 
 static void
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_image.h b/src/venus/vkr_image.h
--- a/src/venus/vkr_image.h
+++ b/src/venus/vkr_image.h
@@ -10,6 +10,22 @@
 
 struct vkr_image {
    struct vkr_object base;
+
+#ifdef __APPLE__
+   /* Set for images the guest created with a DRM modifier. Those are backed by
+    * an optimal-tiled host image, so the host cannot answer a subresource
+    * layout query for them and vkr_dispatch_vkGetImageSubresourceLayout
+    * answers from these instead.
+    */
+   bool drm_modifier;
+   /* every image bound to a memory is kept on that memory's list, so a
+    * scanout read can pick one that matches the picture it wants */
+   struct list_head mem_head;
+   struct vkr_device_memory *bound_mem;
+   uint32_t width;
+   uint32_t height;
+   uint32_t bytes_per_pixel;
+#endif
 };
 VKR_DEFINE_OBJECT_CAST(image, VK_OBJECT_TYPE_IMAGE, VkImage)
 
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_instance.c b/src/venus/vkr_instance.c
--- a/src/venus/vkr_instance.c
+++ b/src/venus/vkr_instance.c
@@ -259,7 +259,9 @@
       return;
    }
 
+   mtx_lock(&ctx->scanout_mutex);
    vkr_instance_destroy(ctx, instance);
+   mtx_unlock(&ctx->scanout_mutex);
 }
 
 void
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_physical_device.c b/src/venus/vkr_physical_device.c
--- a/src/venus/vkr_physical_device.c
+++ b/src/venus/vkr_physical_device.c
@@ -3,6 +3,8 @@
  * SPDX-License-Identifier: MIT
  */
 
+#include <stdio.h>
+
 #include "vkr_physical_device.h"
 
 #include "venus-protocol/vn_protocol_renderer_device.h"
@@ -195,6 +197,77 @@
       physical_dev->gbm_device = vkr_physical_device_get_gbm_device();
 }
 
+#ifdef __APPLE__
+/* MoltenVK reports no dmabuf and no fd-based external memory, so the host
+ * extension list never carries what the guest requires before Mesa will expose
+ * shareable images. Without them zink cannot create a DRI screen and the guest
+ * falls back to llvmpipe, which is what makes a desktop unusable.
+ *
+ * These are advertised to the guest and emulated on top of virtio-gpu blob
+ * resources. vkr_dispatch_vkCreateDevice keeps them from reaching the host,
+ * which would otherwise fail the device create outright.
+ */
+static const char *const vkr_emulated_extensions[] = {
+   "VK_KHR_external_memory_fd",
+   "VK_EXT_external_memory_dma_buf",
+   "VK_EXT_image_drm_format_modifier",
+   "VK_EXT_queue_family_foreign",
+   /* Venus hides VK_KHR_synchronization2 -- and clamps the device to Vulkan
+    * 1.2 -- unless the renderer can import a sync_fd semaphore, which gates
+    * every Vulkan compositor. The host has no sync_fd, so it is emulated on
+    * top of the host's timeline semaphores.
+    */
+   "VK_KHR_external_semaphore_fd",
+};
+
+bool
+vkr_physical_device_extension_is_emulated(const char *name)
+{
+   for (uint32_t i = 0; i < ARRAY_SIZE(vkr_emulated_extensions); i++) {
+      if (!strcmp(name, vkr_emulated_extensions[i]))
+         return true;
+   }
+   return false;
+}
+
+static uint32_t
+vkr_physical_device_add_emulated_extensions(VkExtensionProperties **exts,
+                                            uint32_t count)
+{
+   VkExtensionProperties *grown =
+      realloc(*exts, sizeof(**exts) * (count + ARRAY_SIZE(vkr_emulated_extensions)));
+   if (!grown)
+      return count;
+
+   *exts = grown;
+   for (uint32_t i = 0; i < ARRAY_SIZE(vkr_emulated_extensions); i++) {
+      const char *name = vkr_emulated_extensions[i];
+
+      /* the protocol has to be able to carry it, and the host must not have
+       * advertised it already */
+      if (!vkr_extension_get_spec_version(name))
+         continue;
+
+      bool already_present = false;
+      for (uint32_t j = 0; j < count; j++) {
+         if (!strcmp(grown[j].extensionName, name)) {
+            already_present = true;
+            break;
+         }
+      }
+      if (already_present)
+         continue;
+
+      VkExtensionProperties *props = &grown[count++];
+      memset(props, 0, sizeof(*props));
+      snprintf(props->extensionName, sizeof(props->extensionName), "%s", name);
+      props->specVersion = vkr_extension_get_spec_version(name);
+   }
+
+   return count;
+}
+#endif /* __APPLE__ */
+
 static void
 vkr_physical_device_init_extensions(struct vkr_physical_device *physical_dev,
                                     struct vkr_instance *instance)
@@ -227,8 +300,24 @@
          physical_dev->EXT_external_memory_dma_buf = true;
       else if (!strcmp(props->extensionName, "VK_KHR_external_fence_fd"))
          physical_dev->KHR_external_fence_fd = true;
+#ifdef __APPLE__
+      else if (!strcmp(props->extensionName, "VK_EXT_external_memory_host"))
+         physical_dev->EXT_external_memory_host = true;
+      else if (!strcmp(props->extensionName, "VK_EXT_host_image_copy"))
+         physical_dev->EXT_host_image_copy = true;
+#endif
 
       const uint32_t spec_ver = vkr_extension_get_spec_version(props->extensionName);
+#ifdef VKR_TRACE_COMMANDS
+      {
+         FILE *tf = fopen("/tmp/vkr-extfilter.log", "a");
+         if (tf) {
+            fprintf(tf, "%-50s host_ver=%u spec_ver=%u %s\n", props->extensionName,
+                    props->specVersion, spec_ver, spec_ver ? "KEPT" : "DROPPED");
+            fclose(tf);
+         }
+      }
+#endif
       if (spec_ver) {
          if (props->specVersion > spec_ver)
             props->specVersion = spec_ver;
@@ -253,6 +342,10 @@
          physical_dev->KHR_external_fence_fd = false;
    }
 
+#ifdef __APPLE__
+   advertised_count = vkr_physical_device_add_emulated_extensions(&exts, advertised_count);
+#endif
+
    physical_dev->extensions = exts;
    physical_dev->extension_count = advertised_count;
 }
@@ -543,6 +636,19 @@
 {
    vn_replace_vkGetPhysicalDeviceFeatures2_args_handle(args);
    vkGetPhysicalDeviceFeatures2(args->physicalDevice, args->pFeatures);
+
+#ifdef __APPLE__
+   /* The host hardcodes nullDescriptor to false, and zink refuses to run
+    * without it. The feature is about a descriptor slot being allowed to be
+    * unbound rather than about the values read back, and a correct client never
+    * dereferences such a slot, so report it as supported. It is cleared again
+    * in vkr_dispatch_vkCreateDevice so the host never sees it requested.
+    */
+   VkPhysicalDeviceRobustness2FeaturesEXT *rb2 = vkr_find_struct(
+      args->pFeatures->pNext, VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ROBUSTNESS_2_FEATURES_EXT);
+   if (rb2)
+      rb2->nullDescriptor = VK_TRUE;
+#endif
 }
 
 static void
@@ -585,6 +691,72 @@
    }
 }
 
+#ifdef __APPLE__
+/* An opaque modifier standing for "whatever layout the host picked".
+ *
+ * Linear is not usable here: Apple GPUs cannot render into a linear texture, so
+ * a linear modifier can be sampled from but never be a compositor's output,
+ * which leaves the compositor unable to pick a format for its output at all.
+ * Claiming linear while backing it with an optimal image is worse -- the guest
+ * then asks for the image's subresource layout, which is only valid on linear
+ * images, and the host tears down the VM.
+ *
+ * Describing the layout as opaque is what DRM modifiers are for: the guest
+ * carries the token around and never introspects the bytes, which is exactly
+ * true here, because the pixels are reached through the virtio-gpu resource.
+ */
+/* DRM_FORMAT_MOD_LINEAR.
+ *
+ * The layout the guest is told about, not the one the host uses: the host image
+ * is optimal-tiled because Apple cannot render into a linear texture, and the
+ * layout query is answered from the tightly packed geometry the guest expects.
+ * Linear is the modifier a virtio-gpu scanout plane advertises, and a
+ * compositor picks its output format from the intersection of what the plane
+ * and the renderer offer -- offering anything else leaves that empty.
+ */
+#define VKR_DRM_FORMAT_MOD_HOST_OPAQUE 0ull
+
+/* Report the linear modifier for any format the host can use with linear
+ * tiling. Without this the guest asks for a format's modifier list, the host
+ * ignores the chained struct because MoltenVK does not know it, and every
+ * format comes back with no usable modifier -- which is what stops the guest
+ * from creating a DRI screen and drops the desktop onto llvmpipe.
+ */
+static void
+vkr_physical_device_fill_drm_modifiers(VkFormatProperties2 *props2)
+{
+   VkDrmFormatModifierPropertiesListEXT *list = vkr_find_struct(
+      props2->pNext, VK_STRUCTURE_TYPE_DRM_FORMAT_MODIFIER_PROPERTIES_LIST_EXT);
+   if (!list)
+      return;
+
+
+   /* the modifier is backed by an optimal-tiled host image, so it carries the
+    * optimal features -- notably COLOR_ATTACHMENT, which linear lacks */
+   const VkFormatFeatureFlags feats = props2->formatProperties.optimalTilingFeatures;
+   if (!feats) {
+      list->drmFormatModifierCount = 0;
+      return;
+   }
+
+   if (!list->pDrmFormatModifierProperties) {
+      /* count query */
+      list->drmFormatModifierCount = 1;
+      return;
+   }
+
+   if (!list->drmFormatModifierCount)
+      return;
+
+   list->pDrmFormatModifierProperties[0] = (VkDrmFormatModifierPropertiesEXT){
+      .drmFormatModifier = VKR_DRM_FORMAT_MOD_HOST_OPAQUE,
+      .drmFormatModifierPlaneCount = 1,
+      .drmFormatModifierTilingFeatures = feats,
+   };
+   list->drmFormatModifierCount = 1;
+}
+#endif /* __APPLE__ */
+
 static void
 vkr_dispatch_vkGetPhysicalDeviceFormatProperties2(
    UNUSED struct vn_dispatch_context *dispatch,
@@ -593,6 +765,24 @@
    vn_replace_vkGetPhysicalDeviceFormatProperties2_args_handle(args);
    vkGetPhysicalDeviceFormatProperties2(args->physicalDevice, args->format,
                                         args->pFormatProperties);
+
+#ifdef VKR_TRACE_COMMANDS
+   {
+      const VkFormatProperties *fp = &args->pFormatProperties->formatProperties;
+      const void *chain = args->pFormatProperties->pNext;
+      FILE *f = fopen("/tmp/vkr-fmt.log", "a");
+      if (f) {
+         fprintf(f, "format=%d linear=0x%x optimal=0x%x buffer=0x%x chained=%s\n",
+                 (int)args->format, fp->linearTilingFeatures, fp->optimalTilingFeatures,
+                 fp->bufferFeatures, chain ? "yes" : "no");
+         fclose(f);
+      }
+   }
+#endif
+
+#ifdef __APPLE__
+   vkr_physical_device_fill_drm_modifiers(args->pFormatProperties);
+#endif
 }
 
 static void
@@ -600,9 +790,63 @@
    UNUSED struct vn_dispatch_context *dispatch,
    struct vn_command_vkGetPhysicalDeviceImageFormatProperties2 *args)
 {
+#ifdef __APPLE__
+   /* The host cannot answer a modifier query. Since linear is the only modifier
+    * we advertise, ask the host the equivalent question about linear tiling.
+    */
+   VkPhysicalDeviceImageFormatInfo2 *fmt_info =
+      (VkPhysicalDeviceImageFormatInfo2 *)args->pImageFormatInfo;
+   const VkPhysicalDeviceImageDrmFormatModifierInfoEXT *mod_info = vkr_find_struct(
+      fmt_info->pNext, VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_IMAGE_DRM_FORMAT_MODIFIER_INFO_EXT);
+   VkImageTiling saved_tiling = fmt_info->tiling;
+   if (mod_info) {
+      if (mod_info->drmFormatModifier != VKR_DRM_FORMAT_MOD_HOST_OPAQUE) {
+         args->ret = VK_ERROR_FORMAT_NOT_SUPPORTED;
+         return;
+      }
+      fmt_info->tiling = VK_IMAGE_TILING_OPTIMAL;
+   }
+
+   /* The host does not know the dma_buf handle type and fails the whole query
+    * if it is left in the chain, so unlink it for the duration of the call and
+    * answer that half ourselves below.
+    */
+   VkBaseInStructure *prev_ext_info = vkr_find_prev_struct(
+      fmt_info, VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_EXTERNAL_IMAGE_FORMAT_INFO);
+   VkPhysicalDeviceExternalImageFormatInfo *ext_info =
+      prev_ext_info ? (VkPhysicalDeviceExternalImageFormatInfo *)prev_ext_info->pNext
+                    : NULL;
+   bool emulated_handle = false;
+   if (ext_info &&
+       (ext_info->handleType == VK_EXTERNAL_MEMORY_HANDLE_TYPE_DMA_BUF_BIT_EXT ||
+        ext_info->handleType == VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_FD_BIT)) {
+      emulated_handle = true;
+      prev_ext_info->pNext = ext_info->pNext;
+   }
+#endif
+
    vn_replace_vkGetPhysicalDeviceImageFormatProperties2_args_handle(args);
    args->ret = vkGetPhysicalDeviceImageFormatProperties2(
       args->physicalDevice, args->pImageFormatInfo, args->pImageFormatProperties);
+
+#ifdef __APPLE__
+   if (emulated_handle) {
+      prev_ext_info->pNext = (const struct VkBaseInStructure *)ext_info;
+
+      VkExternalImageFormatProperties *ext_props = vkr_find_struct(
+         args->pImageFormatProperties->pNext,
+         VK_STRUCTURE_TYPE_EXTERNAL_IMAGE_FORMAT_PROPERTIES);
+      if (ext_props && args->ret == VK_SUCCESS) {
+         ext_props->externalMemoryProperties = (VkExternalMemoryProperties){
+            .externalMemoryFeatures = VK_EXTERNAL_MEMORY_FEATURE_IMPORTABLE_BIT |
+                                      VK_EXTERNAL_MEMORY_FEATURE_EXPORTABLE_BIT,
+            .exportFromImportedHandleTypes = ext_info->handleType,
+            .compatibleHandleTypes = ext_info->handleType,
+         };
+      }
+   }
+   fmt_info->tiling = saved_tiling;
+#endif
 }
 
 static void
@@ -630,6 +874,24 @@
    UNUSED struct vn_dispatch_context *dispatch,
    struct vn_command_vkGetPhysicalDeviceExternalSemaphoreProperties *args)
 {
+#ifdef __APPLE__
+   /* The host cannot answer for a sync_fd handle: it has no such handle type.
+    * Report the sync_fd semaphore as importable and exportable, which is what
+    * the guest checks before it will expose VK_KHR_synchronization2.
+    */
+   if (args->pExternalSemaphoreInfo->handleType ==
+       VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_SYNC_FD_BIT) {
+      args->pExternalSemaphoreProperties->exportFromImportedHandleTypes =
+         VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_SYNC_FD_BIT;
+      args->pExternalSemaphoreProperties->compatibleHandleTypes =
+         VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_SYNC_FD_BIT;
+      args->pExternalSemaphoreProperties->externalSemaphoreFeatures =
+         VK_EXTERNAL_SEMAPHORE_FEATURE_EXPORTABLE_BIT |
+         VK_EXTERNAL_SEMAPHORE_FEATURE_IMPORTABLE_BIT;
+      return;
+   }
+#endif
+
    vn_replace_vkGetPhysicalDeviceExternalSemaphoreProperties_args_handle(args);
    vkGetPhysicalDeviceExternalSemaphoreProperties(args->physicalDevice,
                                                   args->pExternalSemaphoreInfo,
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_physical_device.h b/src/venus/vkr_physical_device.h
--- a/src/venus/vkr_physical_device.h
+++ b/src/venus/vkr_physical_device.h
@@ -25,6 +25,14 @@
    bool EXT_external_memory_dma_buf;
 
    bool KHR_external_fence_fd;
+#ifdef __APPLE__
+   /* The host's stand-in for an fd import: memory shared with the guest is
+    * reached through a mapped pointer, and this is what lets the host wrap that
+    * pointer back into a VkDeviceMemory.
+    */
+   bool EXT_external_memory_host;
+   bool EXT_host_image_copy;
+#endif
    bool KHR_external_semaphore_fd;
 
    VkPhysicalDeviceMemoryProperties memory_properties;
@@ -44,4 +52,9 @@
 vkr_physical_device_destroy(struct vkr_context *ctx,
                             struct vkr_physical_device *physical_dev);
 
+#ifdef __APPLE__
+bool
+vkr_physical_device_extension_is_emulated(const char *name);
+#endif
+
 #endif /* VKR_PHYSICAL_DEVICE_H */
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_queue.c b/src/venus/vkr_queue.c
--- a/src/venus/vkr_queue.c
+++ b/src/venus/vkr_queue.c
@@ -479,6 +479,16 @@
 
    vn_replace_vkResetFenceResourceMESA_args_handle(args);
 
+#ifdef __APPLE__
+   /* Both of these only export a sync_fd and close it again, which exists to
+    * flush the payload. The host has no sync_fd and the entry point is not even
+    * loaded, so calling through it kills the VM; ordering is already carried by
+    * the ring, so there is nothing left to do here.
+    */
+   (void)vk;
+   return;
+#endif
+
    const VkFenceGetFdInfoKHR info = {
       .sType = VK_STRUCTURE_TYPE_FENCE_GET_FD_INFO_KHR,
       .fence = args->fence,
@@ -553,6 +563,16 @@
 
    vn_replace_vkWaitSemaphoreResourceMESA_args_handle(args);
 
+#ifdef __APPLE__
+   /* Both of these only export a sync_fd and close it again, which exists to
+    * flush the payload. The host has no sync_fd and the entry point is not even
+    * loaded, so calling through it kills the VM; ordering is already carried by
+    * the ring, so there is nothing left to do here.
+    */
+   (void)vk;
+   return;
+#endif
+
    const VkSemaphoreGetFdInfoKHR info = {
       .sType = VK_STRUCTURE_TYPE_SEMAPHORE_GET_FD_INFO_KHR,
       .semaphore = args->semaphore,
@@ -584,6 +604,33 @@
    /* resourceId 0 is for importing a signaled payload to sync_fd fence */
    assert(!res_info->resourceId);
 
+#ifdef __APPLE__
+   /* The host has no sync_fd, so ImportSemaphoreFdKHR does not exist here and
+    * calling through the null entry would take the VM down. An fd of -1 means
+    * "a payload that is already signalled", so submit nothing and let that
+    * submission signal the semaphore, which leaves it in the state the guest
+    * is about to wait on.
+    */
+   struct vkr_queue *queue = NULL;
+   struct vkr_queue *iter;
+   LIST_FOR_EACH_ENTRY (iter, &dev->queues, base.track_head) {
+      queue = iter;
+      break;
+   }
+   if (!queue) {
+      vkr_context_set_fatal(ctx);
+      return;
+   }
+
+   const VkSubmitInfo submit_info = {
+      .sType = VK_STRUCTURE_TYPE_SUBMIT_INFO,
+      .signalSemaphoreCount = 1,
+      .pSignalSemaphores = &res_info->semaphore,
+   };
+   if (vk->QueueSubmit(queue->base.handle.queue, 1, &submit_info, VK_NULL_HANDLE) !=
+       VK_SUCCESS)
+      vkr_context_set_fatal(ctx);
+#else
    const VkImportSemaphoreFdInfoKHR import_info = {
       .sType = VK_STRUCTURE_TYPE_IMPORT_SEMAPHORE_FD_INFO_KHR,
       .semaphore = res_info->semaphore,
@@ -593,6 +640,7 @@
    };
    if (vk->ImportSemaphoreFdKHR(args->device, &import_info) != VK_SUCCESS)
       vkr_context_set_fatal(ctx);
+#endif
 }
 
 static void
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_renderer.c b/src/venus/vkr_renderer.c
--- a/src/venus/vkr_renderer.c
+++ b/src/venus/vkr_renderer.c
@@ -8,7 +8,18 @@
 #include "venus-protocol/vn_protocol_renderer_info.h"
 #include "virglrenderer_hw.h"
 
+#include <stdio.h>
+
+#include <unistd.h>
+#include <stdlib.h>
+
 #include "vkr_context.h"
+#include "vkr_queue.h"
+#include "vkr_ring.h"
+#include "vkr_device.h"
+#include "vkr_device_memory.h"
+#include "vkr_image.h"
+#include "vkr_physical_device.h"
 
 struct vkr_renderer_state {
    const struct vkr_renderer_callbacks *cbs;
@@ -233,6 +244,320 @@
    return vkr_context_import_resource(ctx, res_id, fd_type, fd, size);
 }
 
+#ifdef __APPLE__
+/* Wait for the guest's rendering to arrive and finish before a scanout read.
+ *
+ * The KMS flush that triggers a scanout read is not gated on the compositor's
+ * render fence here (the virtio-gpu device has no fence passing), so the guest
+ * can request the flush a moment before it has even submitted the frame's
+ * commands to the Venus ring. vkDeviceWaitIdle then finds nothing pending and
+ * the read returns the previous frame. Give an imminent submission a short
+ * window to land, waking the rings so their threads consume it, and wait until
+ * no queue has an unretired fence. A floor covers the not-yet-submitted case;
+ * a cap bounds the cost when nothing is coming.
+ */
+/* Wait for the guest's imminent render to arrive and finish before a scanout
+ * read. The KMS flush that triggers the read is not gated on the compositor's
+ * render fence (the virtio-gpu device has no fence passing), so the guest can
+ * request the flush a moment before it submits the frame's Venus commands, and
+ * vkDeviceWaitIdle then finds nothing to wait for. Watch the rings for a
+ * submission arriving after the flush (their tail advancing), let the ring
+ * threads run it, and wait until it has retired. A short probe window bounds
+ * the wait when nothing is coming; a cap bounds it when something is stuck.
+ */
+static void
+vkr_scanout_wait_ready(void)
+{
+   static uint64_t probe_ns = 5000000;  /* give a late submission this long to appear */
+   static uint64_t cap_ns = 15000000;   /* hard ceiling per read */
+   static int init = 0;
+   if (!init) {
+      init = 1;
+      const char *pr = getenv("SMOLVM_SCANOUT_PROBE_US");
+      const char *cp = getenv("SMOLVM_SCANOUT_CAP_US");
+      if (pr) probe_ns = (uint64_t)strtoull(pr, NULL, 10) * 1000;
+      if (cp) cap_ns = (uint64_t)strtoull(cp, NULL, 10) * 1000;
+   }
+
+   /* Snapshot each ring's tail and wake the ring threads. */
+   uint32_t tails[256];
+   struct vkr_ring *rings[256];
+   unsigned n = 0;
+   list_for_each_entry (struct vkr_context, ctx, &vkr_state.contexts, head) {
+      mtx_lock(&ctx->ring_mutex);
+      list_for_each_entry (struct vkr_ring, ring, &ctx->rings, head) {
+         if (ring->started && n < ARRAY_SIZE(rings)) {
+            rings[n] = ring;
+            tails[n] = vkr_ring_load_tail(ring);
+            n++;
+            vkr_ring_notify(ring);
+         }
+      }
+      mtx_unlock(&ctx->ring_mutex);
+   }
+
+   struct timespec start;
+   clock_gettime(CLOCK_MONOTONIC, &start);
+   bool saw_activity = false;
+   for (;;) {
+      /* A submission that arrived after the flush, work still unconsumed, or a
+       * fence still in flight all count as activity we should wait out. */
+      bool active = false;
+      for (unsigned i = 0; i < n; i++) {
+         const uint32_t tail = vkr_ring_load_tail(rings[i]);
+         if (tail != tails[i] || rings[i]->buffer.cur != tail) {
+            active = true;
+            break;
+         }
+      }
+      if (!active) {
+         list_for_each_entry (struct vkr_context, ctx, &vkr_state.contexts, head) {
+            for (uint32_t i = 0; i < 64 && !active; i++) {
+               struct vkr_queue *q = ctx->sync_queues[i];
+               if (!q)
+                  continue;
+               mtx_lock(&q->sync_thread.mutex);
+               if (!LIST_IS_EMPTY(&q->sync_thread.syncs))
+                  active = true;
+               mtx_unlock(&q->sync_thread.mutex);
+            }
+            if (active)
+               break;
+         }
+      }
+      if (active)
+         saw_activity = true;
+
+      struct timespec now;
+      clock_gettime(CLOCK_MONOTONIC, &now);
+      const uint64_t el = (uint64_t)(now.tv_sec - start.tv_sec) * 1000000000ull +
+                          (now.tv_nsec - start.tv_nsec);
+
+      /* Done once a submission we saw has drained and retired. */
+      if (saw_activity && !active)
+         break;
+      /* Nothing showed up in the probe window: read what we have. */
+      if (!saw_activity && el >= probe_ns)
+         break;
+      if (el >= cap_ns)
+         break;
+
+      const struct timespec nap = { .tv_sec = 0, .tv_nsec = 150000 };
+      nanosleep(&nap, NULL);
+   }
+}
+
+static bool
+vkr_renderer_read_resource_pixels_locked(struct vkr_context *ctx,
+                                         const struct vkr_resource *res,
+                                         uint32_t res_id,
+                                         void *dst,
+                                         uint32_t dst_stride,
+                                         uint32_t width,
+                                         uint32_t height)
+{
+   /* A device teardown frees the memory behind a resource without telling the
+    * resource, so trust the object table rather than the cached pointer.
+    */
+   if (res && res->mem) {
+      mtx_lock(&ctx->object_mutex);
+      const struct hash_entry *live =
+         _mesa_hash_table_search(ctx->object_table, &res->mem_id);
+      const bool ok = live && live->data == res->mem;
+      mtx_unlock(&ctx->object_mutex);
+      if (!ok) {
+         vkr_log("read_resource_pixels: res %u memory is gone", res_id);
+         return false;
+      }
+   }
+   if (!res || !res->mem || !res->mem->bound_image) {
+      vkr_log("read_resource_pixels: res %u has no image bound", res_id);
+      return false;
+   }
+
+   /* The memory may carry several images (a buffer imported more than once,
+    * or a small image reusing an allocation); copying a full frame out of a
+    * smaller one reads past its texture. Take the most recently bound image
+    * that covers the requested picture.
+    */
+   const struct vkr_image *img = NULL;
+   if (res->mem->bound_images.next) {
+      list_for_each_entry_rev (struct vkr_image, cand, &res->mem->bound_images, mem_head) {
+         /* Only an image whose memory is actually bound has host-readable
+          * texture storage; reading an unbound one faults inside Metal. */
+         if (cand->bound_mem && cand->drm_modifier && cand->width >= width &&
+             cand->height >= height) {
+            img = cand;
+            break;
+         }
+      }
+   }
+   if (!img) {
+      vkr_log("read_resource_pixels: res %u has no image covering %ux%u", res_id, width, height);
+      return false;
+   }
+   {
+      mtx_lock(&ctx->object_mutex);
+      const struct hash_entry *live =
+         _mesa_hash_table_search(ctx->object_table, &img->base.id);
+      const bool ok = live && live->data == img;
+      mtx_unlock(&ctx->object_mutex);
+      if (!ok) {
+         vkr_log("read_resource_pixels: res %u image is gone", res_id);
+         return false;
+      }
+   }
+   struct vkr_device *dev = res->mem->device;
+   if (!dev || !dev->physical_device->EXT_host_image_copy)
+      return false;
+
+   PFN_vkrCopyImageToMemory copy_image_to_memory =
+      (PFN_vkrCopyImageToMemory)vkGetDeviceProcAddr(dev->base.handle.device,
+                                                    "vkCopyImageToMemoryEXT");
+   if (!copy_image_to_memory) {
+      copy_image_to_memory = (PFN_vkrCopyImageToMemory)vkGetDeviceProcAddr(
+         dev->base.handle.device, "vkCopyImageToMemory");
+   }
+   if (!copy_image_to_memory) {
+      vkr_log("read_resource_pixels: no host image copy entry point");
+      return false;
+   }
+
+   /* The guest's drawing may still be in flight; a scanout read has no other
+    * ordering against it, and reading early yields an empty picture.
+    */
+   vkr_scanout_wait_ready();
+   vkDeviceWaitIdle(dev->base.handle.device);
+
+
+   const uint32_t bpp = img->bytes_per_pixel ? img->bytes_per_pixel : 4;
+   (void)bpp;
+   const VkrImageToMemoryCopy region = {
+      .sType = VKR_STRUCTURE_TYPE_IMAGE_TO_MEMORY_COPY,
+      .pHostPointer = dst,
+      .memoryRowLength = dst_stride / bpp,
+      .memoryImageHeight = height,
+      .imageSubresource = {
+         .aspectMask = VK_IMAGE_ASPECT_COLOR_BIT,
+         .mipLevel = 0,
+         .baseArrayLayer = 0,
+         .layerCount = 1,
+      },
+      .imageOffset = { 0, 0, 0 },
+      .imageExtent = { width, height, 1 },
+   };
+
+   /* Try the image bound to the resource first, then any other image of the
+    * same size the context owns.
+    *
+    * The guest's driver does not always draw into the image it binds to the
+    * shared allocation: it can keep a private image and hand the shared one a
+    * copy later. Whichever holds the picture is the one worth sending, so take
+    * the first that comes back with something in it.
+    */
+   const VkImageLayout layouts[] = { VK_IMAGE_LAYOUT_GENERAL,
+                                     VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
+                                     VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL };
+
+   struct vkr_image *candidates[64];
+   uint32_t candidate_count = 0;
+   candidates[candidate_count++] = (struct vkr_image *)img;
+
+   const void *scanout_guest_ptr = res->mem ? res->mem->guest_ptr : NULL;
+   mtx_lock(&ctx->object_mutex);
+   hash_table_foreach (ctx->object_table, entry) {
+      struct vkr_object *obj = entry->data;
+      if (candidate_count >= ARRAY_SIZE(candidates))
+         break;
+      if (obj->type != VK_OBJECT_TYPE_IMAGE)
+         continue;
+      struct vkr_image *other = (struct vkr_image *)obj;
+      if (other == img || !other->drm_modifier)
+         continue;
+      /* Skip an image whose memory is not bound: its Metal texture has no
+       * backing yet and copying from it faults. This happens during startup
+       * and teardown, when a re-read can land between image creation and its
+       * memory binding. */
+      if (!other->bound_mem)
+         continue;
+      if (other->width != width || other->height != height)
+         continue;
+      /* Only consider images that wrap the same guest storage as the scanout
+       * resource. A GPU app such as a browser allocates many same-size buffers
+       * of its own; without this they crowd out the desktop's real frame and
+       * the read comes back black. */
+      if (scanout_guest_ptr && other->bound_mem->guest_ptr &&
+          other->bound_mem->guest_ptr != scanout_guest_ptr)
+         continue;
+      candidates[candidate_count++] = other;
+   }
+   mtx_unlock(&ctx->object_mutex);
+
+   bool copied = false;
+   for (uint32_t c = 0; c < candidate_count; c++) {
+      for (uint32_t i = 0; i < ARRAY_SIZE(layouts); i++) {
+         const VkrCopyImageToMemoryInfo info = {
+            .sType = VKR_STRUCTURE_TYPE_COPY_IMAGE_TO_MEMORY_INFO,
+            .srcImage = candidates[c]->base.handle.image,
+            .srcImageLayout = layouts[i],
+            .regionCount = 1,
+            .pRegions = &region,
+         };
+         if (copy_image_to_memory(dev->base.handle.device, &info) != VK_SUCCESS)
+            continue;
+
+         copied = true;
+         const uint8_t *pixels = dst;
+         const size_t bytes = (size_t)dst_stride * height;
+         for (size_t off = 0; off + 3 < bytes; off += 4) {
+            if (pixels[off] | pixels[off + 1] | pixels[off + 2])
+               return true; /* this one holds the picture */
+         }
+         break; /* copied cleanly but empty: try the next image */
+      }
+   }
+
+   if (copied)
+      return true;
+
+   vkr_log("read_resource_pixels: copy failed for res %u", res_id);
+   return false;
+}
+
+bool
+vkr_renderer_read_resource_pixels(uint32_t ctx_id,
+                                  uint32_t res_id,
+                                  void *dst,
+                                  uint32_t dst_stride,
+                                  uint32_t width,
+                                  uint32_t height)
+{
+   /* A scanout is issued on the control queue, which carries no context id, so
+    * find whichever context owns the resource instead of trusting the caller's.
+    */
+   struct vkr_context *ctx = ctx_id ? vkr_renderer_lookup_context(ctx_id) : NULL;
+   const struct vkr_resource *res = ctx ? vkr_context_get_resource(ctx, res_id) : NULL;
+   if (!res) {
+      list_for_each_entry (struct vkr_context, candidate, &vkr_state.contexts, head) {
+         const struct vkr_resource *found = vkr_context_get_resource(candidate, res_id);
+         if (found) {
+            ctx = candidate;
+            res = found;
+            break;
+         }
+      }
+   }
+   if (!ctx || !res) {
+      return false;
+   }
+   mtx_lock(&ctx->scanout_mutex);
+   const bool ok = vkr_renderer_read_resource_pixels_locked(ctx, res, res_id, dst, dst_stride, width, height);
+   mtx_unlock(&ctx->scanout_mutex);
+   return ok;
+}
+
+#endif /* __APPLE__ */
+
 void
 vkr_renderer_destroy_resource(uint32_t ctx_id, uint32_t res_id)
 {
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_renderer.h b/src/venus/vkr_renderer.h
--- a/src/venus/vkr_renderer.h
+++ b/src/venus/vkr_renderer.h
@@ -72,6 +72,16 @@
                              int fd,
                              uint64_t size);
 
+#ifdef __APPLE__
+bool
+vkr_renderer_read_resource_pixels(uint32_t ctx_id,
+                                  uint32_t res_id,
+                                  void *dst,
+                                  uint32_t dst_stride,
+                                  uint32_t width,
+                                  uint32_t height);
+#endif
+
 void
 vkr_renderer_destroy_resource(uint32_t ctx_id, uint32_t res_id);
 
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_ring.c b/src/venus/vkr_ring.c
--- a/src/venus/vkr_ring.c
+++ b/src/venus/vkr_ring.c
@@ -65,7 +65,7 @@
    atomic_store_explicit(ring->control.head, ring_head, memory_order_release);
 }
 
-static uint32_t
+uint32_t
 vkr_ring_load_tail(const struct vkr_ring *ring)
 {
    /* the driver is expected to store the tail with memory_order_release,
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus/vkr_ring.h b/src/venus/vkr_ring.h
--- a/src/venus/vkr_ring.h
+++ b/src/venus/vkr_ring.h
@@ -119,6 +119,9 @@
 void
 vkr_ring_notify(struct vkr_ring *ring);
 
+uint32_t
+vkr_ring_load_tail(const struct vkr_ring *ring);
+
 bool
 vkr_ring_write_extra(struct vkr_ring *ring, size_t offset, uint32_t val);
 
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/venus_context.c b/src/venus_context.c
--- a/src/venus_context.c
+++ b/src/venus_context.c
@@ -1,3 +1,5 @@
+#include <stdio.h>
+
 #include "venus_context.h"
 
 #include "util/hash_table.h"
@@ -7,6 +9,9 @@
 #include "virgl_util.h"
 
 #include "venus/vkr_renderer.h"
+#include "virgl_protocol.h"
+#include "vrend_iov.h"
+#include "pipe/p_state.h"
 
 struct venus_fence {
    uint32_t flags;
@@ -95,12 +100,37 @@
 static int
 venus_context_transfer_3d(struct virgl_context *base,
                           struct virgl_resource *res,
-                          UNUSED const struct vrend_transfer_info *info,
-                          UNUSED int transfer_mode)
+                          const struct vrend_transfer_info *info,
+                          int transfer_mode)
 {
    struct venus_context *ctx = (struct venus_context *)base;
 
-   //fprintf(stderr, "no transfer support for ctx %d and res %d", ctx->base.ctx_id, res->res_id);
+#ifdef __APPLE__
+   /* Reading a rendered image back out.
+    *
+    * The image the guest drew into is optimal-tiled, and on this platform that
+    * means private texture memory -- the bytes are not in any mapping the
+    * display path can read, which is why a scanout came out black. Copy the
+    * image itself into the caller's linear buffer instead.
+    */
+   if (transfer_mode == VIRGL_TRANSFER_FROM_HOST && info && info->iovec_cnt > 0 &&
+       info->iovec && info->box) {
+      const uint32_t width = (uint32_t)info->box->width;
+      const uint32_t height = (uint32_t)info->box->height;
+      const uint32_t stride = info->stride ? info->stride : width * 4;
+      void *dst = info->iovec[0].iov_base;
+
+      if (dst && width && height &&
+          (uint64_t)stride * height <= info->iovec[0].iov_len &&
+          vkr_renderer_read_resource_pixels(ctx->base.ctx_id, res->res_id, dst, stride,
+                                            width, height))
+         return 0;
+   }
+#endif
+
+   (void)ctx;
+   (void)info;
+   (void)transfer_mode;
    return -1;
 }
 
@@ -123,9 +153,10 @@
    uint64_t map_ptr;
    struct virgl_resource_vulkan_info vulkan_info;
 
-   if (!vkr_renderer_create_resource(base->ctx_id, res_id, blob_id, blob_size, blob_flags,
-                                     &fd_type, &res_fd, &map_info, &map_ptr, &vulkan_info)) {
-      //fprintf(stderr, "can't create res %d with blob %llu\n", res_id, blob_id);
+   const bool created =
+      vkr_renderer_create_resource(base->ctx_id, res_id, blob_id, blob_size, blob_flags,
+                                   &fd_type, &res_fd, &map_info, &map_ptr, &vulkan_info);
+   if (!created) {
       return -1;
    }
 
@@ -233,7 +264,14 @@
 static void
 venus_context_retire_fences(struct virgl_context *base)
 {
-   fprintf(stderr, "%s: UNIMPLEMENTED\n", __func__);
+   /* No-op: Venus fences are retired asynchronously through the vkr renderer
+    * callback (venus_state_cb_retire_fence -> venus_context_retire_fences_internal),
+    * which runs whenever a fence signals. The virglrenderer core also invokes
+    * this base callback from its per-poll retire path; doing the work again here
+    * is unnecessary. It must stay silent: the core polls it every frame, so an
+    * fprintf here floods stderr (observed filling the agent log to hundreds of
+    * MB and stalling the GPU worker on the synchronous write). */
+   (void)base;
 #if 0
    struct proxy_context *ctx = (struct proxy_context *)base;
 
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/virglrenderer.c b/src/virglrenderer.c
--- a/src/virglrenderer.c
+++ b/src/virglrenderer.c
@@ -47,6 +47,7 @@
 #include "vrend_winsys.h"
 
 #include "virglrenderer.h"
+#include "venus/vkr_renderer.h"
 #include "virglrenderer_hw.h"
 
 #include "virgl_context.h"
@@ -55,6 +56,7 @@
 
 #include "venus_context.h"
 
+
 struct global_state {
    bool client_initialized;
    void *cookie;
@@ -452,6 +454,25 @@
       vrend_renderer_force_ctx_0();
 }
 
+#ifdef __APPLE__
+/* Read a rendered resource back as linear pixels.
+ *
+ * The display path cannot get at the picture on its own here: the guest renders
+ * into a tiled image that lives in private texture memory, so the blob's
+ * mapping reads as zeroes. This asks the Vulkan renderer for the image itself.
+ */
+int virgl_renderer_read_resource_pixels(uint32_t ctx_id,
+                                        uint32_t res_id,
+                                        void *dst,
+                                        uint32_t stride,
+                                        uint32_t width,
+                                        uint32_t height)
+{
+   return vkr_renderer_read_resource_pixels(ctx_id, res_id, dst, stride, width, height) ? 0
+                                                                                        : -1;
+}
+#endif
+
 void virgl_renderer_ctx_attach_resource(int ctx_id, int res_handle)
 {
    TRACE_FUNC();
diff --color -ruN -x build -x '*.o' -x '*.dylib' -x '*.a' -x .git -x '*.symbols' -x '*.p' a/src/virglrenderer.h b/src/virglrenderer.h
--- a/src/virglrenderer.h
+++ b/src/virglrenderer.h
@@ -311,6 +311,16 @@
 
 VIRGL_EXPORT void virgl_renderer_force_ctx_0(void);
 
+#ifdef __APPLE__
+/* Read a rendered resource back as linear pixels; 0 on success. */
+VIRGL_EXPORT int virgl_renderer_read_resource_pixels(uint32_t ctx_id,
+                                                     uint32_t res_id,
+                                                     void *dst,
+                                                     uint32_t stride,
+                                                     uint32_t width,
+                                                     uint32_t height);
+#endif
+
 VIRGL_EXPORT void virgl_renderer_ctx_attach_resource(int ctx_id, int res_handle);
 VIRGL_EXPORT void virgl_renderer_ctx_detach_resource(int ctx_id, int res_handle);
 
